{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 272,
   "id": "063ba889-7e02-4d63-84b6-4158f6e2b8bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cnv_realign - using detected cnv areas, realign cram file using jump aligner around cnv edges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 280,
   "id": "8b549c0c-9271-49cb-ba29-771e5bbb7f20",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'data/chr22.bed'"
      ]
     },
     "execution_count": 280,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# argumennt processing + debug environment files\n",
    "import sys\n",
    "import os\n",
    "if \"cnv_realign\" in sys.argv[0]:\n",
    "    # commandline invocation\n",
    "    IN_CRAM = sys.argv[1]\n",
    "    CNV_BED = sys.argv[2]\n",
    "    REF_FASTA = sys.argv[3]\n",
    "    OUT_SAM = sys.argv[4]\n",
    "else:\n",
    "    IN_CRAM = \"data/chr22_1M.cram\"\n",
    "    CNV_BED = \"data/chr22.bed\"\n",
    "    REF_FASTA = os.path.expanduser(\"~/tmp/ref/Homo_sapiens_assembly38.fasta\")\n",
    "    OUT_SAM = \"/tmp/chr22_cnv_realign.\" + str(os.getpid()) + \".sam\"\n",
    "\n",
    "FETCH_READ_PADDING = 500\n",
    "FETCH_REF_PADDING = 500\n",
    "\n",
    "OUT_BAM = OUT_SAM.replace(\".sam\", \".bam\")\n",
    "\n",
    "\n",
    "JUMP_ALIGN_CMD = [\"docker\", \"run\", \"-i\", \"jump_align_dev\", \"/bin/bash\", \"-c\", \"jump_align 10 -25 -50 -10 -5 0\"]\n",
    "\n",
    "debug = True;\n",
    "debug_break_after_one_region = False\n",
    "\n",
    "CNV_BED"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 281,
   "id": "ad586376-9e10-4978-9a47-0432160f1e3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# open files\n",
    "import pysam\n",
    "reads_file = pysam.AlignmentFile(IN_CRAM, \"rb\")\n",
    "fasta_file = pysam.FastaFile(REF_FASTA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 282,
   "id": "5e6f5d70-4ee2-41d7-8af7-1e98aa5039d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "\n",
    "def run_process(command, input_path):\n",
    "\n",
    "  with open(input_path, 'r') as file:\n",
    "    input_data = file.read()  \n",
    "    \n",
    "  try:\n",
    "    process = subprocess.Popen(command, \n",
    "                              stdin=subprocess.PIPE, \n",
    "                              stdout=subprocess.PIPE, \n",
    "                              stderr=subprocess.PIPE, \n",
    "                              text=True) \n",
    "    if input_data:\n",
    "      process.stdin.write(input_data)\n",
    "      #process.stdin.close()\n",
    "\n",
    "    stdout, stderr = process.communicate()\n",
    "    returncode = process.returncode\n",
    "    print(stderr)\n",
    "\n",
    "  except subprocess.CalledProcessError as e:\n",
    "    print(f\"Error executing command: {e}\")\n",
    "    return None, e.returncode\n",
    "\n",
    "  return stdout, returncode"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 283,
   "id": "9499a6cc-6b0d-45a4-82ca-7ff9c03b89a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import tempfile\n",
    "\n",
    "# process a single cnv\n",
    "def process_cnv(chrom, start, end, type, tfp):\n",
    "    if debug: print(\"process_cnv: \", locals())\n",
    "\n",
    "    # get all reads that cross the two cnv edges\n",
    "    reads = dict()\n",
    "    refs = []\n",
    "    refs_extents = []\n",
    "    for loc in [int(start), int(end)]:\n",
    "        rmin = loc - FETCH_READ_PADDING\n",
    "        rmax = loc + FETCH_READ_PADDING\n",
    "        for read in reads_file.fetch(chrom, loc - FETCH_READ_PADDING, loc + FETCH_READ_PADDING):\n",
    "            reads[read.qname] = read\n",
    "            rmin = min(rmin, read.reference_start)\n",
    "            rmax = max(rmax, read.reference_end)\n",
    "        refs_extents.append([rmin, rmax])\n",
    "    if debug: print(\"found\", len(reads), \"reads\") \n",
    "    if debug: print(\"refs_extents\", refs_extents)\n",
    "\n",
    "    # get references\n",
    "    for extents in refs_extents:\n",
    "        rmin, rmax = extents\n",
    "        ref = fasta_file.fetch(chrom, rmin, rmax)\n",
    "        refs.append([rmin, ref])\n",
    "    #if debug: print(\"refs\", refs)\n",
    "\n",
    "    # create input file for jump aligner\n",
    "    tmp = \"/tmp/jump_align_input.\" + str(os.getpid())\n",
    "    ref_emitted = False\n",
    "    reads_in_order = []\n",
    "    with open(tmp, 'w') as f:\n",
    "        for read in reads.values():\n",
    "            reads_in_order.append(read)\n",
    "            if not ref_emitted:\n",
    "                line = read.seq + \"\\t\" + refs[0][1] + \"\\t\" + refs[1][1] + \"\\n\"\n",
    "                ref_emitted = True\n",
    "            else:\n",
    "                line = read.seq + \"\\t=\\n\"\n",
    "            f.write(line)\n",
    "    if debug: print(\"tmp\", tmp)\n",
    "\n",
    "    # run jump_align\n",
    "    alignments = run_process(JUMP_ALIGN_CMD, tmp)\n",
    "    header_seen = False\n",
    "    realignments = []\n",
    "    for alignment, read in zip(alignments[0].split(\"\\n\"), [None, *reads_in_order]):\n",
    "        if not header_seen:\n",
    "            header_seen = True;\n",
    "        else:\n",
    "            a = alignment.split(\"\\t\")\n",
    "            realignments.append([read, refs[0][0], refs[1][0], a])\n",
    "    return realignments\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 284,
   "id": "06cfec8e-78fa-41f6-9686-47d19a3ff551",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OUT_SAM /tmp/chr22_cnv_realign.78608.sam\n",
      "process_cnv:  {'chrom': 'chr22', 'start': '22220500', 'end': '22221500', 'type': 'CN6', 'tfp': 'FP'}\n",
      "found 336 reads\n",
      "refs_extents [[22219679, 22221303], [22220666, 22222326]]\n",
      "tmp /tmp/jump_align_input.78608\n",
      "\n",
      "process_cnv:  {'chrom': 'chr22', 'start': '22628000', 'end': '22629000', 'type': 'CN3', 'tfp': 'FP'}\n",
      "found 380 reads\n",
      "refs_extents [[22627186, 22628815], [22628182, 22629798]]\n",
      "tmp /tmp/jump_align_input.78608\n",
      "\n",
      "process_cnv:  {'chrom': 'chr22', 'start': '22630000', 'end': '22631000', 'type': 'CN3', 'tfp': 'FP'}\n",
      "found 323 reads\n",
      "refs_extents [[22629195, 22630829], [22630191, 22631814]]\n",
      "tmp /tmp/jump_align_input.78608\n",
      "\n",
      "process_cnv:  {'chrom': 'chr22', 'start': '22634000', 'end': '22637500', 'type': 'CN4', 'tfp': 'FP'}\n",
      "found 456 reads\n",
      "refs_extents [[22633180, 22634801], [22636682, 22638322]]\n",
      "tmp /tmp/jump_align_input.78608\n",
      "\n",
      "process_cnv:  {'chrom': 'chr22', 'start': '22643500', 'end': '22645000', 'type': 'CN4', 'tfp': 'FP'}\n",
      "found 408 reads\n",
      "refs_extents [[22642675, 22644284], [22644183, 22645846]]\n",
      "tmp /tmp/jump_align_input.78608\n",
      "\n",
      "process_cnv:  {'chrom': 'chr22', 'start': '22652000', 'end': '22653000', 'type': 'CN3', 'tfp': 'FP'}\n",
      "found 313 reads\n",
      "refs_extents [[22651176, 22652762], [22652188, 22653861]]\n",
      "tmp /tmp/jump_align_input.78608\n",
      "\n",
      "process_cnv:  {'chrom': 'chr22', 'start': '22663000', 'end': '22664000', 'type': 'CN3', 'tfp': 'FP'}\n",
      "found 302 reads\n",
      "refs_extents [[22662181, 22663798], [22663156, 22664804]]\n",
      "tmp /tmp/jump_align_input.78608\n",
      "\n",
      "process_cnv:  {'chrom': 'chr22', 'start': '22751000', 'end': '22752000', 'type': 'CN3', 'tfp': 'FP'}\n",
      "found 298 reads\n",
      "refs_extents [[22750202, 22751806], [22751159, 22752832]]\n",
      "tmp /tmp/jump_align_input.78608\n",
      "\n",
      "process_cnv:  {'chrom': 'chr22', 'start': '22762500', 'end': '22763500', 'type': 'CN1', 'tfp': 'TP'}\n",
      "found 186 reads\n",
      "refs_extents [[22761673, 22763293], [22762676, 22764293]]\n",
      "tmp /tmp/jump_align_input.78608\n",
      "\n",
      "process_cnv:  {'chrom': 'chr22', 'start': '22765000', 'end': '22768000', 'type': 'CN1', 'tfp': 'TP'}\n",
      "found 176 reads\n",
      "refs_extents [[22764189, 22765686], [22767213, 22768796]]\n",
      "tmp /tmp/jump_align_input.78608\n",
      "\n",
      "process_cnv:  {'chrom': 'chr22', 'start': '22771500', 'end': '22773000', 'type': 'CN1', 'tfp': 'TP'}\n",
      "found 186 reads\n",
      "refs_extents [[22770683, 22772332], [22772179, 22773834]]\n",
      "tmp /tmp/jump_align_input.78608\n",
      "\n",
      "process_cnv:  {'chrom': 'chr22', 'start': '22793500', 'end': '22794500', 'type': 'CN1', 'tfp': 'TP'}\n",
      "found 183 reads\n",
      "refs_extents [[22792708, 22794335], [22793705, 22795323]]\n",
      "tmp /tmp/jump_align_input.78608\n",
      "\n",
      "process_cnv:  {'chrom': 'chr22', 'start': '22821500', 'end': '22822500', 'type': 'CN1', 'tfp': 'TP'}\n",
      "found 168 reads\n",
      "refs_extents [[22820680, 22822343], [22821673, 22823305]]\n",
      "tmp /tmp/jump_align_input.78608\n",
      "\n",
      "process_cnv:  {'chrom': 'chr22', 'start': '22855000', 'end': '22856000', 'type': 'CN1', 'tfp': 'TP'}\n",
      "found 168 reads\n",
      "refs_extents [[22854180, 22855846], [22855204, 22856800]]\n",
      "tmp /tmp/jump_align_input.78608\n",
      "\n",
      "process_cnv:  {'chrom': 'chr22', 'start': '22891000', 'end': '22892000', 'type': 'CN1', 'tfp': 'TP'}\n",
      "found 178 reads\n",
      "refs_extents [[22890247, 22891809], [22891154, 22892802]]\n",
      "tmp /tmp/jump_align_input.78608\n",
      "\n",
      "process_cnv:  {'chrom': 'chr22', 'start': '22893500', 'end': '22895000', 'type': 'CN1', 'tfp': 'TP'}\n",
      "found 201 reads\n",
      "refs_extents [[22892673, 22894268], [22894179, 22895825]]\n",
      "tmp /tmp/jump_align_input.78608\n",
      "\n",
      "process_cnv:  {'chrom': 'chr22', 'start': '22896000', 'end': '22897500', 'type': 'CN1', 'tfp': 'TP'}\n",
      "found 194 reads\n",
      "refs_extents [[22895190, 22896792], [22896674, 22898348]]\n",
      "tmp /tmp/jump_align_input.78608\n",
      "\n",
      "process_cnv:  {'chrom': 'chr22', 'start': '22900000', 'end': '22905500', 'type': 'CN1', 'tfp': 'TP'}\n",
      "found 281 reads\n",
      "refs_extents [[22899186, 22900820], [22904667, 22906310]]\n",
      "tmp /tmp/jump_align_input.78608\n",
      "\n",
      "cmd samtools sort /tmp/chr22_cnv_realign.78608.sam >/tmp/chr22_cnv_realign.78608.bam\n",
      "cmd samtools index /tmp/chr22_cnv_realign.78608.bam\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 284,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# open output sam file\n",
    "print(\"OUT_SAM\", OUT_SAM)\n",
    "sam_file = pysam.AlignmentFile(OUT_SAM, \"w\", header=reads_file.header)\n",
    "\n",
    "debug_qname = \"036742_1-Z0027-1118527563\"\n",
    "\n",
    "# loop on bed file\n",
    "with open(CNV_BED) as f:\n",
    "    for line in f:\n",
    "        bed_line = line.strip().split()\n",
    "        realignments = process_cnv(*bed_line)\n",
    "        for realignment in realignments:\n",
    "            read, ref1_start, ref2_start, ainfo = realignment\n",
    "            if read.qname == debug_qname:\n",
    "                print(\"ref_start1/2\", ref1_start, ref2_start)\n",
    "                print(\"ainfo\", ainfo)\n",
    "\n",
    "            # write original read\n",
    "            tags = read.tags\n",
    "            read.tags = tags + [('JT', 0)]\n",
    "\n",
    "            sam_file.write(read)\n",
    "\n",
    "            # decode alignment info\n",
    "            score, jumpInsertSize, jumpRange, \\\n",
    "                jbegin1, jcigar1, jreadlen1, jreflen1, \\\n",
    "                jbegin2, jcigar2, jreadlen2, jreflen2, \\\n",
    "                score1, begin1, cigar1, readlen1, reflen1, \\\n",
    "                score2, begin2, cigar2, readlen2, reflen2 = ainfo\n",
    "\n",
    "            # build read aligned to references\n",
    "            qname = read.qname\n",
    "            for i in [1,2]:\n",
    "                read.qname = qname + \"_REF\" + str(i)\n",
    "                read.tags = tags + [('JT', i)]    \n",
    "                if i == 1:\n",
    "                    read.cigarstring = cigar1\n",
    "                    read.reference_start = ref1_start + int(begin1)\n",
    "                    read.tags += [('JS', int(score1))]\n",
    "                else:\n",
    "                    read.cigarstring = cigar2\n",
    "                    read.reference_start = ref2_start + int(begin2)\n",
    "                    read.tags += [('JS', int(score2))]\n",
    "                sam_file.write(read)\n",
    "                \n",
    "            # build read aligned to both references with a jump in the middle\n",
    "            if len(jcigar1) and len(jcigar2):\n",
    "                end_on_ref1 = ref1_start + int(jbegin1) + int(jreflen1)\n",
    "                start_on_ref2 = ref2_start + int(jbegin2)\n",
    "                delta = start_on_ref2 - end_on_ref1\n",
    "                read.qname = qname + \"_JUMP\"\n",
    "                new_cigar = jcigar1\n",
    "                if int(jumpInsertSize):\n",
    "                    new_cigar += jumpInsertSize + \"I\"\n",
    "                new_cigar += str(delta) + \"D\"\n",
    "                new_cigar += jcigar2\n",
    "                read.cigarstring = new_cigar\n",
    "                read.reference_start = ref1_start + int(jbegin1)\n",
    "                read.tags = tags + [('JT', 3)]   \n",
    "                read.tags += [('JS', int(score))]\n",
    "                sam_file.write(read)  \n",
    "                if qname == debug_qname:\n",
    "                    print(\"jump new_cigar\", new_cigar)\n",
    "\n",
    "        if debug_break_after_one_region:\n",
    "            break\n",
    "\n",
    "sam_file.close()\n",
    "\n",
    "# convert to sorted bam\n",
    "cmd = \"samtools sort \" + OUT_SAM + \" >\" + OUT_BAM\n",
    "print(\"cmd\", cmd)\n",
    "os.system(cmd)\n",
    "cmd = \"samtools index \" + OUT_BAM\n",
    "print(\"cmd\", cmd)\n",
    "os.system(cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a512c2fb-716d-4664-9694-ff0ad5fdf610",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa847017-4121-4a57-894e-749f2d1fbe50",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b87fea7-8524-49c0-84c7-c2ea4dffe448",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
