{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "063ba889-7e02-4d63-84b6-4158f6e2b8bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cnv_realign - using detected cnv areas, realign cram file using jump aligner around cnv edges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "8b549c0c-9271-49cb-ba29-771e5bbb7f20",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['/Users/drorkessler/miniconda3/lib/python3.10/site-packages/ipykernel_launcher.py', '-f', '/Users/drorkessler/Library/Jupyter/runtime/kernel-beb565a3-1980-4bdd-b972-5cd8a0dc98fd.json']\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'data/chr22_fp1.bed'"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# argumennt processing + debug environment files\n",
    "import sys\n",
    "import os\n",
    "\n",
    "MATCH_SCORE = 2\n",
    "\n",
    "print(sys.argv)\n",
    "tmp = \"/tmp/jump_align_input.\" + str(os.getpid())\n",
    "JUMP_ALIGN_CMD = (\"jump_align/jump_align \" + str(MATCH_SCORE) + \" -8 -18 -1 -1 0 \" + tmp).split(\" \")\n",
    "if \"cnv_realign\" in sys.argv[0] or \"stdin\" in sys.argv[0]:\n",
    "    # commandline invocation\n",
    "    IN_CRAM = sys.argv[1]\n",
    "    CNV_BED = sys.argv[2]\n",
    "    REF_FASTA = sys.argv[3]\n",
    "    OUT_SAM = sys.argv[4] + \".sam\"\n",
    "else:\n",
    "    IN_CRAM = \"data/chr22_1M.cram\"\n",
    "    IN_CRAM = os.path.expanduser(\"~/tmp/data/jump_align/chr22.cram\")\n",
    "    CNV_BED = \"data/chr22_fp1.bed\"\n",
    "    REF_FASTA = os.path.expanduser(\"~/tmp/ref/Homo_sapiens_assembly38.fasta\")\n",
    "    OUT_SAM = \"/tmp/chr22_cnv_realign.\" + str(os.getpid()) + \".sam\"\n",
    "\n",
    "\n",
    "# debug overrides\n",
    "# IN_CRAM = os.path.expanduser(\"~/tmp/data/jump_align/chr22.cram\")\n",
    "# CNV_BED = \"data/chr22_ex3.bed\"\n",
    "\n",
    "\n",
    "FETCH_READ_PADDING = 500\n",
    "FETCH_REF_PADDING = 0\n",
    "MIN_SEQ_LEN_JUMP_ALIGN_COMPONENT = 30\n",
    "MIN_GAP_LEN = 30\n",
    "MAX_READS_PER_CNV = 2000\n",
    "\n",
    "JUMP_SCORE_THRESHOLD = 0.85\n",
    "\n",
    "OUT_BAM = OUT_SAM.replace(\".sam\", \".bam\")\n",
    "OUT_BED = OUT_SAM.replace(\".sam\", \".bed\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "debug = True;\n",
    "debug_break_after_one_region = False\n",
    "debug_break_after_debug_seen = False\n",
    "debug_output_debug_only = False\n",
    "\n",
    "\n",
    "\n",
    "CNV_BED"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "ad586376-9e10-4978-9a47-0432160f1e3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# open files\n",
    "import pysam\n",
    "reads_file = pysam.AlignmentFile(IN_CRAM, \"rb\", reference_filename=REF_FASTA)\n",
    "fasta_file = pysam.FastaFile(REF_FASTA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "5e6f5d70-4ee2-41d7-8af7-1e98aa5039d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "\n",
    "def run_process(command, input_path):\n",
    "  command \n",
    "  print(\"command\", command) \n",
    "    \n",
    "  try:\n",
    "    process = subprocess.Popen(command, \n",
    "                              stdin=subprocess.PIPE, \n",
    "                              stdout=subprocess.PIPE, \n",
    "                              stderr=subprocess.PIPE, \n",
    "                              text=True) \n",
    "\n",
    "    stdout, stderr = process.communicate()\n",
    "    returncode = process.returncode\n",
    "    #print(stderr)\n",
    "\n",
    "  except subprocess.CalledProcessError as e:\n",
    "    print(f\"Error executing command: {e}\")\n",
    "    return None, e.returncode\n",
    "\n",
    "  return stdout, returncode"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "9499a6cc-6b0d-45a4-82ca-7ff9c03b89a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import tempfile\n",
    "import random\n",
    "\n",
    "random.seed(0)\n",
    "\n",
    "# process a single cnv\n",
    "def process_cnv(chrom, start, end, type, tfp):\n",
    "    if debug: print(\"process_cnv: \", locals())\n",
    "\n",
    "    # get all reads that cross the two cnv edges\n",
    "    reads = dict()\n",
    "    reads_in_ref = [set(), set()]\n",
    "    refs = []\n",
    "    refs_extents = []\n",
    "    ref_id = 0\n",
    "    for loc in [int(start), int(end)]:\n",
    "        rmin = loc - FETCH_READ_PADDING\n",
    "        rmax = loc + FETCH_READ_PADDING\n",
    "        for read in reads_file.fetch(chrom, loc - FETCH_READ_PADDING, loc + FETCH_READ_PADDING):\n",
    "            reads[read.qname] = read\n",
    "            rmin = min(rmin, read.reference_start)\n",
    "            rmax = max(rmax, read.reference_end)\n",
    "            reads_in_ref[ref_id].add(read.qname)\n",
    "        refs_extents.append([rmin, rmax])\n",
    "        ref_id += 1\n",
    "    # extend references before and after\n",
    "    refs_extents[0][0] = max(0, refs_extents[0][0] - FETCH_REF_PADDING)\n",
    "    refs_extents[1][1] = refs_extents[1][1] + FETCH_REF_PADDING\n",
    "    if debug: print(\"found\", len(reads), \"reads\") \n",
    "    if debug: print(\"refs_extents\", refs_extents)\n",
    "\n",
    "    # get references\n",
    "    for extents in refs_extents:\n",
    "        rmin, rmax = extents\n",
    "        ref = fasta_file.fetch(chrom, rmin, rmax)\n",
    "        refs.append([rmin, ref])\n",
    "    #if debug: print(\"refs\", refs)\n",
    "\n",
    "    # create input file for jump aligner\n",
    "    ref_emitted = False\n",
    "    reads_in_order = []\n",
    "    subsample_ratio = 1.0\n",
    "    if len(reads) > MAX_READS_PER_CNV:\n",
    "        subsample_ratio = MAX_READS_PER_CNV / len(reads)\n",
    "        print(\"subsample_ratio\", subsample_ratio)\n",
    "    with open(tmp, 'w') as f:\n",
    "        for read in reads.values():\n",
    "            if subsample_ratio < 1.0:\n",
    "                if random.random() > subsample_ratio:\n",
    "                    continue\n",
    "            reads_in_order.append(read)\n",
    "            if not ref_emitted:\n",
    "                line = read.qname + \"\\t\" + read.seq + \"\\t\" + refs[0][1] + \"\\t\" + refs[1][1] + \"\\n\"\n",
    "                ref_emitted = True\n",
    "            else:\n",
    "                line = read.qname + \"\\t\" + read.seq + \"\\t=\\n\"\n",
    "            f.write(line)\n",
    "    if debug: print(\"tmp\", tmp)\n",
    "\n",
    "    # run jump_align\n",
    "    alignments = run_process(JUMP_ALIGN_CMD, tmp)\n",
    "    header_seen = False\n",
    "    realignments = []\n",
    "    for alignment, read in zip(alignments[0].split(\"\\n\"), [None, *reads_in_order]):\n",
    "        if not header_seen:\n",
    "            header_seen = True;\n",
    "        else:\n",
    "            a = alignment.split(\"\\t\")\n",
    "            in1 = read.qname in reads_in_ref[0]\n",
    "            in2 = read.qname in reads_in_ref[1]\n",
    "            realignments.append([read, refs[0][0], refs[1][0], a, in1, in2])\n",
    "    return realignments\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "06cfec8e-78fa-41f6-9686-47d19a3ff551",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OUT_SAM /tmp/chr22_cnv_realign.9224.sam\n",
      "process_cnv:  {'chrom': 'chr22', 'start': '10679500', 'end': '10687000', 'type': 'CN1', 'tfp': 'FP'}\n",
      "found 452 reads\n",
      "refs_extents [[10678750, 10680267], [10686239, 10687805]]\n",
      "tmp /tmp/jump_align_input.9224\n",
      "command ['jump_align/jump_align', '2', '-8', '-18', '-1', '-1', '0', '/tmp/jump_align_input.9224']\n",
      "ref_start1/2 10678750 10686239\n",
      "ainfo ['036742_1-Z0027-0297955940', '519', '20', '0', '359', '14=1X11=1X46=5D13=1X21=1X16=1X48=1X43=1X26=1X50=', '296', '301', '284', '23=1X3=1X7=', '35', '35', '419', '359', '14=1X11=1X46=5D13=1X21=1X16=1X48=1X43=1X26=1X50=54I1=', '351', '302', '-279', '281', '2=314I1D23=1X3=1X7=', '351', '38']\n",
      "in_ref [True, False]\n",
      "036742_1-Z0027-0297955940 ideal_score 702 ideal_ratio 0.7393162393162394 score 519\n",
      "jump new_cigar 14=1X11=1X46=5D13=1X21=1X16=1X48=1X43=1X26=1X50=20I7113D23=1X3=1X7=\n",
      "036742_1-Z0027-2326442142 ideal_score 712 ideal_ratio 0.7429775280898876 score 529\n",
      "036742_1-Z0027-3266931725 ideal_score 700 ideal_ratio 0.7385714285714285 score 517\n",
      "036742_2-Z0027-1077014883 ideal_score 624 ideal_ratio 0.7387820512820513 score 461\n",
      "036742_2-Z0027-2431418746 ideal_score 678 ideal_ratio 0.7640117994100295 score 518\n",
      "036742_1-Z0027-2182702108 ideal_score 564 ideal_ratio 0.75177304964539 score 424\n",
      "036742_2-Z0027-0828146711 ideal_score 624 ideal_ratio 0.7596153846153846 score 474\n",
      "036742_2-Z0027-3117995689 ideal_score 616 ideal_ratio 0.7564935064935064 score 466\n",
      "036742_1-Z0027-3534515816 ideal_score 660 ideal_ratio 0.7575757575757576 score 500\n",
      "036742_1-Z0027-2406794556 ideal_score 504 ideal_ratio 0.7023809523809523 score 354\n",
      "036742_2-Z0027-1629083607 ideal_score 682 ideal_ratio 0.6744868035190615 score 460\n",
      "036742_2-Z0027-2227099255 ideal_score 686 ideal_ratio 0.7813411078717201 score 536\n",
      "036742_1-Z0027-0002430687 ideal_score 682 ideal_ratio 0.5953079178885631 score 406\n",
      "036742_2-Z0027-2360676293 ideal_score 676 ideal_ratio 0.7781065088757396 score 526\n",
      "036742_1-Z0027-2739209447 ideal_score 512 ideal_ratio 0.70703125 score 362\n",
      "036742_1-Z0027-0943826711 ideal_score 686 ideal_ratio 0.7667638483965015 score 526\n",
      "036742_1-Z0027-3804017552 ideal_score 654 ideal_ratio 0.7706422018348624 score 504\n",
      "036742_2-Z0027-3508424741 ideal_score 532 ideal_ratio 0.45112781954887216 score 240\n",
      "036742_1-Z0027-2746145392 ideal_score 650 ideal_ratio 0.7692307692307693 score 500\n",
      "036742_2-Z0027-0907874447 ideal_score 570 ideal_ratio 0.7368421052631579 score 420\n",
      "036742_2-Z0027-0197012603 ideal_score 480 ideal_ratio 0.7291666666666666 score 350\n",
      "036742_1-Z0027-0941595903 ideal_score 568 ideal_ratio 0.7359154929577465 score 418\n",
      "036742_1-Z0027-3623632040 ideal_score 634 ideal_ratio 0.6608832807570978 score 419\n",
      "036742_2-Z0027-1986075951 ideal_score 680 ideal_ratio 0.7941176470588235 score 540\n",
      "036742_2-Z0027-3195906365 ideal_score 684 ideal_ratio 0.7807017543859649 score 534\n",
      "036742_2-Z0027-1902969424 ideal_score 624 ideal_ratio 0.7548076923076923 score 471\n",
      "036742_2-Z0027-2548033969 ideal_score 534 ideal_ratio 0.6423220973782772 score 343\n",
      "036742_2-Z0027-2019891658 ideal_score 682 ideal_ratio 0.5850439882697948 score 399\n",
      "036742_1-Z0027-2967076001 ideal_score 654 ideal_ratio 0.753822629969419 score 493\n",
      "036742_2-Z0027-0003583933 ideal_score 490 ideal_ratio 0.7142857142857143 score 350\n",
      "036742_2-Z0027-0837047005 ideal_score 684 ideal_ratio 0.72953216374269 score 499\n",
      "036742_1-Z0027-2775423028 ideal_score 600 ideal_ratio 0.7666666666666667 score 460\n",
      "036742_2-Z0027-2889454149 ideal_score 676 ideal_ratio 0.7485207100591716 score 506\n",
      "036742_2-Z0027-3456737882 ideal_score 516 ideal_ratio 0.748062015503876 score 386\n",
      "036742_2-Z0027-0518255118 ideal_score 668 ideal_ratio 0.7440119760479041 score 497\n",
      "036742_2-Z0027-0695938620 ideal_score 386 ideal_ratio 0.7150259067357513 score 276\n",
      "036742_1-Z0027-2121693072 ideal_score 666 ideal_ratio 0.5900900900900901 score 393\n",
      "036742_2-Z0027-1255571050 ideal_score 632 ideal_ratio 0.6693037974683544 score 423\n",
      "036742_1-Z0027-0862555141 ideal_score 640 ideal_ratio 0.571875 score 366\n",
      "036742_1-Z0027-2192583873 ideal_score 674 ideal_ratio 0.8071216617210683 score 544\n",
      "036742_1-Z0027-0647302353 ideal_score 546 ideal_ratio 0.7619047619047619 score 416\n",
      "036742_1-Z0027-0174766945 ideal_score 680 ideal_ratio 0.8088235294117647 score 550\n",
      "036742_2-Z0027-2163627121 ideal_score 688 ideal_ratio 0.811046511627907 score 558\n",
      "036742_2-Z0027-3156173626 ideal_score 692 ideal_ratio 0.7976878612716763 score 552\n",
      "036742_1-Z0027-1172338998 ideal_score 632 ideal_ratio 0.7642405063291139 score 483\n",
      "036742_2-Z0027-0194056332 ideal_score 680 ideal_ratio 0.6308823529411764 score 429\n",
      "036742_1-Z0027-2815047844 ideal_score 636 ideal_ratio 0.779874213836478 score 496\n",
      "036742_2-Z0027-0619394991 ideal_score 496 ideal_ratio 0.5604838709677419 score 278\n",
      "036742_1-Z0027-0513045687 ideal_score 656 ideal_ratio 0.801829268292683 score 526\n",
      "jump_read_written 48 jump_read_lowqual 1\n",
      "cmd samtools sort /tmp/chr22_cnv_realign.9224.sam >/tmp/chr22_cnv_realign.9224.bam\n",
      "cmd samtools index /tmp/chr22_cnv_realign.9224.bam\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# open output sam file\n",
    "print(\"OUT_SAM\", OUT_SAM)\n",
    "sam_file = pysam.AlignmentFile(OUT_SAM, \"w\", header=reads_file.header)\n",
    "\n",
    "debug_qname = \"036742_1-Z0027-0297955940\"\n",
    "debug_seen = False\n",
    "\n",
    "# loop on bed file, write output bed file\n",
    "with open(OUT_BED, \"w\") as out_bed:\n",
    "    with open(CNV_BED) as f:\n",
    "        for line in f:\n",
    "            bed_line = line.strip().split()\n",
    "            realignments = process_cnv(*bed_line)\n",
    "            jump_read_written = 0\n",
    "            jump_read_lowqual = 0\n",
    "            for realignment in realignments:\n",
    "                in_ref = [False, False]\n",
    "                read, ref1_start, ref2_start, ainfo, in_ref[0], in_ref[1] = realignment\n",
    "                if read.qname == debug_qname:\n",
    "                    print(\"ref_start1/2\", ref1_start, ref2_start)\n",
    "                    print(\"ainfo\", ainfo)\n",
    "                    print(\"in_ref\", in_ref)\n",
    "                    debug_seen = True\n",
    "    \n",
    "                # write original read\n",
    "                tags = read.tags\n",
    "                read.tags = tags + [('JT', 0)]\n",
    "    \n",
    "                if not debug_output_debug_only or read.qname == debug_qname:\n",
    "                    sam_file.write(read)\n",
    "    \n",
    "                # decode alignment info\n",
    "                qname1, score, jumpInsertSize, jumpRange, \\\n",
    "                    jbegin1, jcigar1, jreadlen1, jreflen1, \\\n",
    "                    jbegin2, jcigar2, jreadlen2, jreflen2, \\\n",
    "                    score1, begin1, cigar1, readlen1, reflen1, \\\n",
    "                    score2, begin2, cigar2, readlen2, reflen2 = ainfo\n",
    "    \n",
    "                # build read aligned to references\n",
    "                qname = read.qname\n",
    "                if qname != qname1:\n",
    "                    printf(\"qname mismatch!\", qname, qname1)\n",
    "                for i in [1,2]:\n",
    "                    if not in_ref[i-1]:\n",
    "                        continue\n",
    "                    read.qname = qname + \"_REF\" + str(i)\n",
    "                    read.tags = tags + [('JT', i)]    \n",
    "                    if i == 1:\n",
    "                        read.cigarstring = cigar1\n",
    "                        read.reference_start = ref1_start + int(begin1)\n",
    "                        read.tags += [('JS', int(score1))]\n",
    "                    else:\n",
    "                        read.cigarstring = cigar2\n",
    "                        read.reference_start = ref2_start + int(begin2)\n",
    "                        read.tags += [('JS', int(score2))]\n",
    "                    if not debug_output_debug_only or qname == debug_qname:    \n",
    "                        sam_file.write(read)\n",
    "\n",
    "                # build read aligned to both references with a jump in the middle\n",
    "                if len(jcigar1) and len(jcigar2) \\\n",
    "                        and int(jreadlen1) >= MIN_SEQ_LEN_JUMP_ALIGN_COMPONENT \\\n",
    "                        and int(jreadlen2) >= MIN_SEQ_LEN_JUMP_ALIGN_COMPONENT \\\n",
    "                        and int(jreflen1) >= MIN_SEQ_LEN_JUMP_ALIGN_COMPONENT \\\n",
    "                        and int(jreflen2) >= MIN_SEQ_LEN_JUMP_ALIGN_COMPONENT:\n",
    "\n",
    "                    # calculate ratio of jump score to idle score \n",
    "                    ideal_score = len(read.seq) * MATCH_SCORE\n",
    "                    ideal_ratio = int(score) / ideal_score\n",
    "                    print(qname, \"ideal_score\", ideal_score, \"ideal_ratio\", ideal_ratio, \"score\", score)\n",
    "                    if ideal_ratio >= JUMP_SCORE_THRESHOLD:\n",
    "                        end_on_ref1 = ref1_start + int(jbegin1) + int(jreflen1)\n",
    "                        start_on_ref2 = ref2_start + int(jbegin2)\n",
    "                        delta = start_on_ref2 - end_on_ref1\n",
    "                        read.qname = qname + \"_JUMP\"\n",
    "                        new_cigar = jcigar1\n",
    "                        if int(jumpInsertSize):\n",
    "                            if False:\n",
    "                                delta -= int(jumpInsertSize)\n",
    "                            else:\n",
    "                                new_cigar += jumpInsertSize + \"I\"\n",
    "                        if delta > 0:\n",
    "                            new_cigar += str(delta) + \"D\"\n",
    "                        if delta >= MIN_GAP_LEN:\n",
    "                            new_cigar += jcigar2\n",
    "                            read.cigarstring = new_cigar\n",
    "                            read.reference_start = ref1_start + int(jbegin1)\n",
    "                            read.tags = tags + [('JT', 3)]   \n",
    "                            read.tags += [('JS', int(score))]\n",
    "                            if not debug_output_debug_only or qname == debug_qname:    \n",
    "                                sam_file.write(read)  \n",
    "                                jump_read_written += 1\n",
    "                            if qname == debug_qname:\n",
    "                                print(\"jump new_cigar\", new_cigar)\n",
    "                    else:\n",
    "                        jump_read_lowqual += 1\n",
    "    \n",
    "            if debug: print(\"jump_read_written\", jump_read_written, \"jump_read_lowqual\", jump_read_lowqual)\n",
    "            out_bed.write(line[:-1] + \"\\t\" + str(jump_read_written) + \"\\t\" + str(jump_read_lowqual) + \"\\n\")\n",
    "            \n",
    "            if debug_break_after_one_region:\n",
    "                break\n",
    "            if debug_break_after_debug_seen and debug_seen:\n",
    "                break\n",
    "\n",
    "sam_file.close()\n",
    "\n",
    "# convert to sorted bam\n",
    "cmd = \"samtools sort \" + OUT_SAM + \" >\" + OUT_BAM\n",
    "print(\"cmd\", cmd)\n",
    "os.system(cmd)\n",
    "cmd = \"samtools index \" + OUT_BAM\n",
    "print(\"cmd\", cmd)\n",
    "os.system(cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a512c2fb-716d-4664-9694-ff0ad5fdf610",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa847017-4121-4a57-894e-749f2d1fbe50",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b87fea7-8524-49c0-84c7-c2ea4dffe448",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0c0fa51-0880-46f6-98ac-2be37f66e863",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
