{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "063ba889-7e02-4d63-84b6-4158f6e2b8bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cnv_realign - using detected cnv areas, realign cram file using jump aligner around cnv edges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "8b549c0c-9271-49cb-ba29-771e5bbb7f20",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['/Users/drorkessler/miniconda3/lib/python3.10/site-packages/ipykernel_launcher.py', '-f', '/Users/drorkessler/Library/Jupyter/runtime/kernel-76dca9ec-b520-4f7e-be35-d69bcad3d7d8.json']\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'/Users/drorkessler/tmp/data/jump_align/chr1_1.bed'"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# argumennt processing + debug environment files\n",
    "import sys\n",
    "import os\n",
    "\n",
    "MATCH_SCORE = 2\n",
    "MISMATCH_SCORE = -8\n",
    "OPEN_SCORE = -18\n",
    "EXTEND_SCORE = -1\n",
    "JUMP_SCORE = 0\n",
    "MODE=\"DEL\"\n",
    "\n",
    "MIN_MISMATCHES = 0\n",
    "\n",
    "DEBUG_DUP = False\n",
    "\n",
    "print(sys.argv)\n",
    "tmp = \"/tmp/jump_align_input.\" + str(os.getpid())\n",
    "JUMP_ALIGN_CMD = [\"jump_align\", str(MATCH_SCORE), str(MISMATCH_SCORE), str(OPEN_SCORE), str(EXTEND_SCORE), \"-1\", str(JUMP_SCORE), tmp]\n",
    "if \"cnv_realign\" in sys.argv[0] or \"stdin\" in sys.argv[0]:\n",
    "    if len(sys.argv) < 5:\n",
    "        print(\"usage: \" + sys.argv[0] + \" <input-cram> <range-bed> <ref-fasta> <output-prefix> [<mode>] [<min-mismatches]\\n\")\n",
    "        sys.exit(-1)\n",
    "    # commandline invocation\n",
    "    IN_CRAM = sys.argv[1]\n",
    "    CNV_BED = sys.argv[2]\n",
    "    REF_FASTA = sys.argv[3]\n",
    "    OUT_SAM = sys.argv[4] + \".sam\"\n",
    "    if len(sys.argv) >= 6:\n",
    "        MODE = sys.argv[5]\n",
    "    if len(sys.argv) >= 7:\n",
    "        MIN_MISMATCHES = int(sys.argv[6])\n",
    "else:\n",
    "    #IN_CRAM = os.path.expanduser(\"~/tmp/data/jump_align/chr12_1.cram\")\n",
    "    #CNV_BED = os.path.expanduser(\"~/tmp/data/jump_align/chr12_1.bed\")\n",
    "    IN_CRAM = os.path.expanduser(\"~/tmp/data/jump_align/chr1.cram\")\n",
    "    CNV_BED = os.path.expanduser(\"~/tmp/data/jump_align/chr1_1.bed\")\n",
    "    \n",
    "    #IN_CRAM = os.path.expanduser(\"~/tmp/data/jump_align/NA24385-Z0027.cram\")\n",
    "    #CNV_BED = os.path.expanduser(\"~/tmp/data/jump_align/chr17_1.bed\")\n",
    "    \n",
    "    #IN_CRAM = os.path.expanduser(\"~/tmp/data/jump_align/250218_Ilya/036686-Sample_5-Z0005-CATGTATCCTCTGAT_chr17_1.bam\")\n",
    "    #CNV_BED = os.path.expanduser(\"~/tmp/data/jump_align/250218_Ilya/chr17_1.bed\")\n",
    "    \n",
    "    \n",
    "    REF_FASTA = os.path.expanduser(\"~/tmp/ref/Homo_sapiens_assembly38.fasta\")\n",
    "    OUT_SAM = \"/tmp/cnv_realign.\" + str(os.getpid()) + \".sam\"\n",
    "    MODE = \"DUP\"\n",
    "\n",
    "    # pair haps test\n",
    "    #IN_CRAM = os.path.expanduser(\"~/tmp/data/pair_haps/0030945-Z0114_merged_assembly.bam\")\n",
    "    #CNV_BED = os.path.expanduser(\"~/tmp/data/pair_haps/chr4_ex2.bed\")\n",
    "    #MODE = \"DEL\"\n",
    "\n",
    "    #IN_CRAM = \"/tmp/s1/pair_haps.69480.bam\"\n",
    "    #CNV_BED = \"/tmp/s1/haps.bed\"\n",
    "    MODE = \"DEL_HAPS\"\n",
    "    JUMP_ALIGN_CMD = [\"jump_align/jump_align\"] + JUMP_ALIGN_CMD[1:]\n",
    "\n",
    "    MIN_MISMATCHES = 30\n",
    "\n",
    "JT_INCLUDE = []\n",
    "QNAME_JUMP = True\n",
    "if MODE == \"DEL_HAPS\":\n",
    "    JT_INCLUDE = [3, 100, 150]\n",
    "    MODE = \"DEL\"\n",
    "    QNAME_JUMP = False\n",
    "\n",
    "\n",
    "FETCH_READ_PADDING = 500\n",
    "FETCH_REF_PADDING = 0\n",
    "MIN_SEQ_LEN_JUMP_ALIGN_COMPONENT = 30\n",
    "MIN_GAP_LEN = 30\n",
    "MAX_READS_PER_CNV = 4000\n",
    "\n",
    "JUMP_SCORE_THRESHOLD = 0.85\n",
    "REF_SCORE_THRESHOLD = 1.0\n",
    "\n",
    "OUT_BAM = OUT_SAM.replace(\".sam\", \".bam\")\n",
    "OUT_BED = OUT_SAM.replace(\".sam\", \".bed\")\n",
    "\n",
    "debug = True;\n",
    "debug_break_after_one_region = False\n",
    "debug_break_after_debug_seen = False\n",
    "debug_output_debug_only = False\n",
    "\n",
    "CNV_BED"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "ad586376-9e10-4978-9a47-0432160f1e3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# open files\n",
    "import pysam\n",
    "reads_file = pysam.AlignmentFile(IN_CRAM, \"rb\", reference_filename=REF_FASTA)\n",
    "fasta_file = pysam.FastaFile(REF_FASTA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "5e6f5d70-4ee2-41d7-8af7-1e98aa5039d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "\n",
    "def run_process(command, input_path):\n",
    "  command \n",
    "  print(\"command\", command) \n",
    "    \n",
    "  try:\n",
    "    process = subprocess.Popen(command, \n",
    "                              stdin=subprocess.PIPE, \n",
    "                              stdout=subprocess.PIPE, \n",
    "                              stderr=subprocess.PIPE, \n",
    "                              text=True) \n",
    "\n",
    "    stdout, stderr = process.communicate()\n",
    "    returncode = process.returncode\n",
    "    #print(stderr)\n",
    "\n",
    "  except subprocess.CalledProcessError as e:\n",
    "    print(f\"Error executing command: {e}\")\n",
    "    return None, e.returncode\n",
    "\n",
    "  return stdout, returncode"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "8904d4bc-a80d-4e5f-9356-9889cf619106",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "def count_md_mismatches(read):\n",
    "    \"\"\"\n",
    "    Count the number of mismatches in a pysam.AlignedSegment read using the MD tag.\n",
    "    Indels are not counted.\n",
    "    \"\"\"\n",
    "    try:\n",
    "        md_tag = read.get_tag(\"MD\")\n",
    "    except KeyError:\n",
    "        return None\n",
    "\n",
    "    # Find all letters in the MD string, which represent mismatches\n",
    "    mismatches = re.findall(r\"[A-Z]\", md_tag)\n",
    "    return len(mismatches)\n",
    "\n",
    "def count_nm_mismatches(read):\n",
    "    \"\"\"\n",
    "    Count the number of mismatches in a pysam.AlignedSegment read using the NM tag.\n",
    "    Indels are counted.\n",
    "    \"\"\"\n",
    "    try:\n",
    "        nm_tag = read.get_tag(\"NM\")\n",
    "    except KeyError:\n",
    "        return None\n",
    "\n",
    "    return int(nm_tag)\n",
    "\n",
    "def count_softclip_mismatches(read, reference):\n",
    "    \"\"\"\n",
    "    Count mismatches in soft-clipped regions (both left and right) of a read.\n",
    "    `read` is a pysam.AlignedSegment\n",
    "    `reference` is a pysam.FastaFile\n",
    "    \"\"\"\n",
    "    if read.is_unmapped:\n",
    "        return 0\n",
    "\n",
    "    seq = read.query_sequence\n",
    "    mismatches = 0\n",
    "    ref_name = read.reference_name\n",
    "    start = read.reference_start\n",
    "    end = read.reference_end\n",
    "\n",
    "    cigartuples = read.cigartuples\n",
    "    # CIGAR operations\n",
    "    SOFT_CLIP = 4\n",
    "\n",
    "    # Left soft clip\n",
    "    if cigartuples[0][0] == SOFT_CLIP:\n",
    "        clip_len = cigartuples[0][1]\n",
    "        clipped_bases = seq[:clip_len]\n",
    "        ref_start = max(0, start - clip_len)\n",
    "        ref_bases = reference.fetch(ref_name, ref_start, start)\n",
    "        for rb, qb in zip(ref_bases, clipped_bases):\n",
    "            if rb.upper() != qb.upper():\n",
    "                mismatches += 1\n",
    "\n",
    "    # Right soft clip\n",
    "    if cigartuples[-1][0] == SOFT_CLIP:\n",
    "        clip_len = cigartuples[-1][1]\n",
    "        clipped_bases = seq[-clip_len:]\n",
    "        ref_bases = reference.fetch(ref_name, end, end + clip_len)\n",
    "        for rb, qb in zip(ref_bases, clipped_bases):\n",
    "            if rb.upper() != qb.upper():\n",
    "                mismatches += 1\n",
    "\n",
    "    return mismatches\n",
    "\n",
    "md_values = []\n",
    "nm_values = []\n",
    "sc_values = []\n",
    "\n",
    "def accept_read(read):\n",
    "    if MIN_MISMATCHES <= 0:\n",
    "        return True\n",
    "    sc = count_softclip_mismatches(read, fasta_file);\n",
    "    nm = count_nm_mismatches(read)\n",
    "    return (sc + nm) >= MIN_MISMATCHES\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "9499a6cc-6b0d-45a4-82ca-7ff9c03b89a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import tempfile\n",
    "import random\n",
    "\n",
    "random.seed(0)\n",
    "\n",
    "# is read soft clipped\n",
    "def is_softclipped(read):\n",
    "    return read.cigartuples[0][0] == pysam.CSOFT_CLIP or read.cigartuples[-1][0] == pysam.CSOFT_CLIP\n",
    "\n",
    "def is_substential_softclipped(read):\n",
    "    return (read.cigartuples[0][0] == pysam.CSOFT_CLIP and read.cigartuples[0][1]) >= 30 \\\n",
    "                or (read.cigartuples[-1][0] == pysam.CSOFT_CLIP and read.cigartuples[-1][1] == 30)\n",
    "\n",
    "# process a single cnv\n",
    "def process_cnv(chrom, start, end, mode):\n",
    "    if debug: print(\"process_cnv: \", locals())\n",
    "\n",
    "    # get all reads that cross the two cnv edges\n",
    "    reads = dict()\n",
    "    reads_in_ref = [set(), set()]\n",
    "    refs = []\n",
    "    refs_extents = []\n",
    "    ref_id = 0\n",
    "    for loc in [start, end]:\n",
    "        rmin = max(0, loc - FETCH_READ_PADDING)\n",
    "        rmax = loc + FETCH_READ_PADDING\n",
    "        for read in reads_file.fetch(chrom, max(0, loc - FETCH_READ_PADDING), loc + FETCH_READ_PADDING):\n",
    "            if mode == \"DUP\" and not is_substential_softclipped(read):\n",
    "                continue\n",
    "            if mode == \"DUP\" and read.is_supplementary:\n",
    "                continue\n",
    "            reads[read.qname] = read\n",
    "            rmin = min(rmin, read.reference_start)\n",
    "            rmax = max(rmax, read.reference_end)\n",
    "            reads_in_ref[ref_id].add(read.qname)\n",
    "        refs_extents.append([rmin, rmax])\n",
    "        ref_id += 1\n",
    "        \n",
    "    # extend references before and after\n",
    "    refs_extents[0][0] = max(0, refs_extents[0][0] - FETCH_REF_PADDING)\n",
    "    refs_extents[1][1] = refs_extents[1][1] + FETCH_REF_PADDING\n",
    "    if debug: print(\"found\", len(reads), \"reads\") \n",
    "    if debug: print(\"refs_extents\", refs_extents)\n",
    "\n",
    "    # get references\n",
    "    for extents in refs_extents:\n",
    "        rmin, rmax = extents\n",
    "        ref = fasta_file.fetch(chrom, rmin, rmax)\n",
    "        refs.append([rmin, ref])\n",
    "    #if debug: print(\"refs\", refs)\n",
    "\n",
    "    # create input file for jump aligner\n",
    "    ref_emitted = False\n",
    "    reads_in_order = []\n",
    "    subsample_ratio = 1.0\n",
    "    if len(reads) > MAX_READS_PER_CNV:\n",
    "        subsample_ratio = MAX_READS_PER_CNV / len(reads)\n",
    "        print(\"subsample_ratio\", subsample_ratio)\n",
    "    nlines = 0\n",
    "    with open(tmp, 'w') as f:\n",
    "        for read in reads.values():\n",
    "            if subsample_ratio < 1.0:\n",
    "                if random.random() > subsample_ratio:\n",
    "                    continue\n",
    "            if not accept_read(read):\n",
    "                continue\n",
    "            reads_in_order.append(read)\n",
    "            if not ref_emitted:\n",
    "                line = read.qname + \"\\t\" + read.seq + \"\\t\" + refs[0][1] + \"\\t\" + refs[1][1] + \"\\n\"\n",
    "                if mode == \"DUP\":\n",
    "                    line = read.qname + \"\\t\" + read.seq + \"\\t\" + refs[1][1] + \"\\t\" + refs[0][1] + \"\\n\"\n",
    "                ref_emitted = True\n",
    "            else:\n",
    "                line = read.qname + \"\\t\" + read.seq + \"\\t=\\n\"\n",
    "            f.write(line)\n",
    "            nlines += 1\n",
    "    if debug: print(\"tmp\", tmp)\n",
    "\n",
    "    # run jump_align\n",
    "    alignments = run_process(JUMP_ALIGN_CMD, tmp)\n",
    "    header_seen = False\n",
    "    realignments = []\n",
    "    rheader = []\n",
    "    for alignment, read in zip(alignments[0].split(\"\\n\"), [None, *reads_in_order]):\n",
    "        if not header_seen:\n",
    "            rheader = alignment.split(\"\\t\")\n",
    "            header_seen = True;\n",
    "        else:\n",
    "            a = alignment.split(\"\\t\")\n",
    "            in1 = read.qname in reads_in_ref[0]\n",
    "            in2 = read.qname in reads_in_ref[1]\n",
    "            realignments.append([read, refs[0][0], refs[1][0], a, in1, in2])\n",
    "    print(\"nlines\", nlines)\n",
    "    return (rheader, realignments, nlines)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "6f7fe1a2-2012-4801-be85-fa397bc3cbf8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def sized_substr(s, l):\n",
    "    return [s[ofs:ofs+l] for ofs in range(len(s) - l + 1)]\n",
    "\n",
    "def sized_dict(s, l):\n",
    "    d = {}\n",
    "    for ss in sized_substr(s, l):\n",
    "        if ss in d:\n",
    "            d[ss] += 1\n",
    "        else:\n",
    "            d[ss] = 1\n",
    "    return d\n",
    "\n",
    "def sized_list(s, l):\n",
    "    sl = [(x,y) for x,y in sized_dict(s, l).items()]\n",
    "    return sorted(sl, key=lambda x: x[1], reverse=True)\n",
    "\n",
    "    return sl\n",
    "\n",
    "def top_sized(s):\n",
    "    ts = [sized_list(s,x)[0] for x in range(2, len(s)-2)]\n",
    "    return ts\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "06cfec8e-78fa-41f6-9686-47d19a3ff551",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OUT_SAM /tmp/cnv_realign.65969.sam\n",
      "process_cnv:  {'chrom': 'chr1', 'start': 16500, 'end': 17500, 'mode': 'DEL'}\n",
      "found 805 reads\n",
      "refs_extents [[15663, 17320], [16673, 18344]]\n",
      "tmp /tmp/jump_align_input.65969\n",
      "command ['jump_align/jump_align', '2', '-8', '-18', '-1', '-1', '0', '/tmp/jump_align_input.65969']\n",
      "nlines 15\n",
      "jump_read_written 0 jump_read_lowscore 15 jump_read_lessthanref 11\n",
      "process_cnv:  {'chrom': 'chr1', 'start': 189000, 'end': 190000, 'mode': 'DEL'}\n",
      "found 640 reads\n",
      "refs_extents [[188172, 189823], [189200, 190829]]\n",
      "tmp /tmp/jump_align_input.65969\n",
      "command ['jump_align/jump_align', '2', '-8', '-18', '-1', '-1', '0', '/tmp/jump_align_input.65969']\n",
      "nlines 8\n",
      "jump_read_written 0 jump_read_lowscore 8 jump_read_lessthanref 6\n",
      "process_cnv:  {'chrom': 'chr1', 'start': 190500, 'end': 192000, 'mode': 'DEL'}\n",
      "found 792 reads\n",
      "refs_extents [[189687, 191337], [191194, 192846]]\n",
      "tmp /tmp/jump_align_input.65969\n",
      "command ['jump_align/jump_align', '2', '-8', '-18', '-1', '-1', '0', '/tmp/jump_align_input.65969']\n",
      "nlines 15\n",
      "jump_read_written 0 jump_read_lowscore 15 jump_read_lessthanref 6\n",
      "process_cnv:  {'chrom': 'chr1', 'start': 269500, 'end': 271000, 'mode': 'DEL'}\n",
      "found 601 reads\n",
      "refs_extents [[268701, 270336], [270171, 271828]]\n",
      "tmp /tmp/jump_align_input.65969\n",
      "command ['jump_align/jump_align', '2', '-8', '-18', '-1', '-1', '0', '/tmp/jump_align_input.65969']\n",
      "nlines 10\n",
      "jump_read_written 0 jump_read_lowscore 10 jump_read_lessthanref 3\n",
      "process_cnv:  {'chrom': 'chr1', 'start': 434000, 'end': 435000, 'mode': 'DEL'}\n",
      "found 242 reads\n",
      "refs_extents [[433195, 434706], [434275, 435784]]\n",
      "tmp /tmp/jump_align_input.65969\n",
      "command ['jump_align/jump_align', '2', '-8', '-18', '-1', '-1', '0', '/tmp/jump_align_input.65969']\n",
      "nlines 34\n",
      "jump_read_written 0 jump_read_lowscore 34 jump_read_lessthanref 17\n",
      "process_cnv:  {'chrom': 'chr1', 'start': 595000, 'end': 596000, 'mode': 'DEL'}\n",
      "found 358 reads\n",
      "refs_extents [[594179, 595804], [595186, 596755]]\n",
      "tmp /tmp/jump_align_input.65969\n",
      "command ['jump_align/jump_align', '2', '-8', '-18', '-1', '-1', '0', '/tmp/jump_align_input.65969']\n",
      "nlines 6\n",
      "jump_read_written 0 jump_read_lowscore 6 jump_read_lessthanref 5\n",
      "process_cnv:  {'chrom': 'chr1', 'start': 606500, 'end': 608000, 'mode': 'DEL'}\n",
      "found 331 reads\n",
      "refs_extents [[605676, 607326], [607172, 608837]]\n",
      "tmp /tmp/jump_align_input.65969\n",
      "command ['jump_align/jump_align', '2', '-8', '-18', '-1', '-1', '0', '/tmp/jump_align_input.65969']\n",
      "nlines 5\n",
      "jump_read_written 0 jump_read_lowscore 5 jump_read_lessthanref 3\n",
      "process_cnv:  {'chrom': 'chr1', 'start': 613500, 'end': 614500, 'mode': 'DEL'}\n",
      "found 266 reads\n",
      "refs_extents [[612675, 614300], [613688, 615327]]\n",
      "tmp /tmp/jump_align_input.65969\n",
      "command ['jump_align/jump_align', '2', '-8', '-18', '-1', '-1', '0', '/tmp/jump_align_input.65969']\n",
      "nlines 3\n",
      "jump_read_written 0 jump_read_lowscore 3 jump_read_lessthanref 1\n",
      "process_cnv:  {'chrom': 'chr1', 'start': 615000, 'end': 616000, 'mode': 'DEL'}\n",
      "found 284 reads\n",
      "refs_extents [[614145, 615828], [615168, 616809]]\n",
      "tmp /tmp/jump_align_input.65969\n",
      "command ['jump_align/jump_align', '2', '-8', '-18', '-1', '-1', '0', '/tmp/jump_align_input.65969']\n",
      "nlines 4\n",
      "jump_read_written 0 jump_read_lowscore 4 jump_read_lessthanref 4\n",
      "process_cnv:  {'chrom': 'chr1', 'start': 625500, 'end': 627000, 'mode': 'DEL'}\n",
      "found 377 reads\n",
      "refs_extents [[624693, 626321], [626169, 627745]]\n",
      "tmp /tmp/jump_align_input.65969\n",
      "command ['jump_align/jump_align', '2', '-8', '-18', '-1', '-1', '0', '/tmp/jump_align_input.65969']\n",
      "nlines 13\n",
      "jump_read_written 0 jump_read_lowscore 13 jump_read_lessthanref 4\n",
      "process_cnv:  {'chrom': 'chr1', 'start': 669500, 'end': 670500, 'mode': 'DEL'}\n",
      "found 295 reads\n",
      "refs_extents [[668758, 670328], [669715, 671305]]\n",
      "tmp /tmp/jump_align_input.65969\n",
      "command ['jump_align/jump_align', '2', '-8', '-18', '-1', '-1', '0', '/tmp/jump_align_input.65969']\n",
      "nlines 31\n",
      "jump_read_written 0 jump_read_lowscore 30 jump_read_lessthanref 9\n",
      "process_cnv:  {'chrom': 'chr1', 'start': 673000, 'end': 675500, 'mode': 'DEL'}\n",
      "found 425 reads\n",
      "refs_extents [[672182, 673820], [674707, 676337]]\n",
      "tmp /tmp/jump_align_input.65969\n",
      "command ['jump_align/jump_align', '2', '-8', '-18', '-1', '-1', '0', '/tmp/jump_align_input.65969']\n",
      "nlines 7\n",
      "jump_read_written 0 jump_read_lowscore 7 jump_read_lessthanref 1\n",
      "process_cnv:  {'chrom': 'chr1', 'start': 1729500, 'end': 1730500, 'mode': 'DEL'}\n",
      "found 397 reads\n",
      "refs_extents [[1728703, 1730316], [1729710, 1731280]]\n",
      "tmp /tmp/jump_align_input.65969\n",
      "command ['jump_align/jump_align', '2', '-8', '-18', '-1', '-1', '0', '/tmp/jump_align_input.65969']\n",
      "nlines 6\n",
      "jump_read_written 0 jump_read_lowscore 6 jump_read_lessthanref 3\n",
      "process_cnv:  {'chrom': 'chr1', 'start': 2658000, 'end': 2659000, 'mode': 'DEL'}\n",
      "found 1027 reads\n",
      "refs_extents [[2657213, 2658843], [2658289, 2659794]]\n",
      "tmp /tmp/jump_align_input.65969\n",
      "command ['jump_align/jump_align', '2', '-8', '-18', '-1', '-1', '0', '/tmp/jump_align_input.65969']\n",
      "nlines 223\n",
      "jump_read_written 1 jump_read_lowscore 222 jump_read_lessthanref 50\n",
      "process_cnv:  {'chrom': 'chr1', 'start': 2682000, 'end': 2683000, 'mode': 'DEL'}\n",
      "found 994 reads\n",
      "refs_extents [[2681204, 2682799], [2682202, 2683782]]\n",
      "tmp /tmp/jump_align_input.65969\n",
      "command ['jump_align/jump_align', '2', '-8', '-18', '-1', '-1', '0', '/tmp/jump_align_input.65969']\n",
      "nlines 278\n",
      "jump_read_written 1 jump_read_lowscore 277 jump_read_lessthanref 40\n",
      "process_cnv:  {'chrom': 'chr1', 'start': 2683500, 'end': 2685000, 'mode': 'DEL'}\n",
      "found 999 reads\n",
      "refs_extents [[2682683, 2684316], [2684199, 2685796]]\n",
      "tmp /tmp/jump_align_input.65969\n",
      "command ['jump_align/jump_align', '2', '-8', '-18', '-1', '-1', '0', '/tmp/jump_align_input.65969']\n",
      "nlines 429\n",
      "jump_read_written 0 jump_read_lowscore 428 jump_read_lessthanref 64\n",
      "process_cnv:  {'chrom': 'chr1', 'start': 2686000, 'end': 2688500, 'mode': 'DEL'}\n",
      "found 850 reads\n",
      "refs_extents [[2685248, 2686807], [2687696, 2689295]]\n",
      "tmp /tmp/jump_align_input.65969\n",
      "command ['jump_align/jump_align', '2', '-8', '-18', '-1', '-1', '0', '/tmp/jump_align_input.65969']\n",
      "nlines 401\n",
      "jump_read_written 0 jump_read_lowscore 401 jump_read_lessthanref 54\n",
      "cmd samtools sort /tmp/cnv_realign.65969.sam >/tmp/cnv_realign.65969.bam\n",
      "cmd samtools index /tmp/cnv_realign.65969.bam\n",
      "nlines_total 1488\n",
      "CPU times: user 624 ms, sys: 93.6 ms, total: 717 ms\n",
      "Wall time: 12.5 s\n"
     ]
    }
   ],
   "source": [
    "import statistics\n",
    "\n",
    "# open output sam file\n",
    "print(\"OUT_SAM\", OUT_SAM)\n",
    "sam_file = pysam.AlignmentFile(OUT_SAM, \"w\", header=reads_file.header)\n",
    "\n",
    "debug_qname = \"036686_2-Z0005-3658287334\"\n",
    "debug_seen = False\n",
    "\n",
    "def dump_dup_read(f, read, ainfo, better):\n",
    "    f.write(\"%s\\t%d\\t%s\\t%s\\n%s\\n\" % (read.qname, read.reference_start, read.cigarstring, str(better), str(ainfo)))\n",
    "\n",
    "chrom_sizes = dict(zip(fasta_file.references, fasta_file.lengths))\n",
    "# loop on bed file, write output bed file\n",
    "nlines_total = 0\n",
    "with open(OUT_BED, \"w\") as out_bed:\n",
    "    with open(CNV_BED) as f:\n",
    "        for line in f:\n",
    "            if line.startswith(\"#\"):\n",
    "                continue\n",
    "            bed_line = line.strip().split()\n",
    "            bed_chrom, bed_start, bed_end = bed_line[:3]\n",
    "            bed_start = int(bed_start)\n",
    "            bed_end = int(bed_end)\n",
    "            #check for valid cnv\n",
    "            if bed_end + FETCH_READ_PADDING > chrom_sizes[bed_chrom]:\n",
    "                continue\n",
    "            rheader, realignments, nlines = process_cnv(bed_chrom, bed_start, bed_end, MODE)\n",
    "            nlines_total += nlines\n",
    "            jump_read_written = 0\n",
    "            jump_read_lowscore = 0\n",
    "            jump_read_okscore = 0\n",
    "            jump_read_lessthanref = 0\n",
    "            jump_better = 0\n",
    "            jdelsize = []\n",
    "            jjumpland = []\n",
    "            jump_better_delta = []\n",
    "            for realignment in realignments:\n",
    "                in_ref = [False, False]\n",
    "                read, ref1_start, ref2_start, ainfo, in_ref[0], in_ref[1] = realignment\n",
    "                if read.qname == debug_qname:\n",
    "                    print(\"ref_start1/2\", ref1_start, ref2_start)\n",
    "                    print(\"in_ref\", in_ref)\n",
    "                    for e in [(x,y) for x,y in zip(rheader, ainfo)]:\n",
    "                        print(\"%s: %s\" % e)\n",
    "                    debug_seen = True\n",
    "    \n",
    "                # write original read\n",
    "                tags = read.tags\n",
    "                read.tags = tags + [('JT', 0)]\n",
    "    \n",
    "                if not debug_output_debug_only or read.qname == debug_qname:\n",
    "                    if not len(JT_INCLUDE) or 0 in JT_INCLUDE:\n",
    "                        sam_file.write(read)\n",
    "    \n",
    "                # decode alignment info\n",
    "                qname1, score, jumpInsertSize, jumpRange, \\\n",
    "                    jbegin1, jcigar1, jreadlen1, jreflen1, \\\n",
    "                    jbegin2, jcigar2, jreadlen2, jreflen2, \\\n",
    "                    score1, begin1, cigar1, readlen1, reflen1, \\\n",
    "                    score2, begin2, cigar2, readlen2, reflen2 = ainfo\n",
    "                score = int(score)\n",
    "                score1 = int(score1)\n",
    "                score2 = int(score2)\n",
    "\n",
    "                # jump score better?\n",
    "                better = False\n",
    "                if score > 0 and score > max(score1, score2):\n",
    "                    if int(jreadlen1) >= MIN_SEQ_LEN_JUMP_ALIGN_COMPONENT \\\n",
    "                        and int(jreadlen2) >= MIN_SEQ_LEN_JUMP_ALIGN_COMPONENT \\\n",
    "                        and int(jreflen1) >= MIN_SEQ_LEN_JUMP_ALIGN_COMPONENT \\\n",
    "                        and int(jreflen2) >= MIN_SEQ_LEN_JUMP_ALIGN_COMPONENT:\n",
    "                        jump_better += 1\n",
    "                        jump_better_delta.append(score - max(score1, score2))\n",
    "                        better = True\n",
    "                        if DEBUG_DUP:\n",
    "                            dump_dup_read(out_bed, read, ainfo, better)\n",
    "    \n",
    "                # build read aligned to references\n",
    "                qname = read.qname\n",
    "                reference_start = read.reference_start\n",
    "                if qname == debug_qname: print(\"reference_start\", reference_start)\n",
    "                ref_score = -1000000\n",
    "                if qname != qname1:\n",
    "                    printf(\"qname mismatch!\", qname, qname1)\n",
    "                for i in [1,2]:\n",
    "                    if not in_ref[i-1]:\n",
    "                        continue\n",
    "                    read.qname = qname + \"_REF\" + str(i)\n",
    "                    read.tags = tags + [('JT', i)]    \n",
    "                    if i == 1:\n",
    "                        read.cigarstring = cigar1\n",
    "                        read.reference_start = ref1_start + int(begin1)\n",
    "                        read.tags += [('JS', int(score1))]\n",
    "                        ref_score = max(ref_score, int(score1))\n",
    "                    else:\n",
    "                        read.cigarstring = cigar2\n",
    "                        read.reference_start = ref2_start + int(begin2)\n",
    "                        read.tags += [('JS', int(score2))]\n",
    "                        ref_score = max(ref_score, int(score2))\n",
    "                    if not debug_output_debug_only or qname == debug_qname:    \n",
    "                        if not len(JT_INCLUDE) or i in JT_INCLUDE:\n",
    "                            sam_file.write(read)\n",
    "\n",
    "                # adjust jump score to include jumpInsertSize\n",
    "                jump_score = int(score)\n",
    "                if int(jumpInsertSize):\n",
    "                    jump_score += (OPEN_SCORE + int(jumpInsertSize) * EXTEND_SCORE)\n",
    "\n",
    "                # calculate ratio of jump score to idle score \n",
    "                ideal_score = len(read.seq) * MATCH_SCORE\n",
    "                ideal_ratio = int(jump_score) / ideal_score\n",
    "                #print(qname, \"ideal_score\", ideal_score, \"ideal_ratio\", ideal_ratio, \"jump_score\", jump_score)\n",
    "                if ideal_ratio < JUMP_SCORE_THRESHOLD:\n",
    "                    jump_read_lowscore += 1\n",
    "                else:\n",
    "                    jump_read_okscore += 1\n",
    "\n",
    "                # DUP JT=3\n",
    "                if MODE == \"DUP\" and better:\n",
    "                    read.tags = tags + [('JT', 3)]   \n",
    "                    read.tags += [('JS', score)]\n",
    "                    if QNAME_JUMP:\n",
    "                        read.qname = qname + \"_JUMP\"\n",
    "                    else:\n",
    "                        read.qname = qname\n",
    "                    if not len(JT_INCLUDE) or 3 in JT_INCLUDE:\n",
    "                        sam_file.write(read)\n",
    "                        \n",
    "                # build read aligned to both references with a jump in the middle\n",
    "                if MODE == \"DEL\" and len(jcigar1) and len(jcigar2) :\n",
    "\n",
    "                    jt = 0\n",
    "                    if int(jreadlen1) < MIN_SEQ_LEN_JUMP_ALIGN_COMPONENT \\\n",
    "                        or int(jreadlen2) < MIN_SEQ_LEN_JUMP_ALIGN_COMPONENT \\\n",
    "                        or int(jreflen1) < MIN_SEQ_LEN_JUMP_ALIGN_COMPONENT \\\n",
    "                        or int(jreflen2) < MIN_SEQ_LEN_JUMP_ALIGN_COMPONENT:\n",
    "                        jt += 50\n",
    "\n",
    "                    if ideal_ratio < JUMP_SCORE_THRESHOLD:\n",
    "                        jt += 100\n",
    "                    if jump_score <= (ref_score * REF_SCORE_THRESHOLD):\n",
    "                        jt += 200\n",
    "                        jump_read_lessthanref += 1                        \n",
    "                    end_on_ref1 = ref1_start + int(jbegin1) + int(jreflen1)\n",
    "                    start_on_ref2 = ref2_start + int(jbegin2)\n",
    "                    land = (start_on_ref2 - bed_start) / (bed_end - bed_start)\n",
    "                    delta = start_on_ref2 - end_on_ref1\n",
    "                    if QNAME_JUMP:\n",
    "                        read.qname = qname + \"_JUMP\"\n",
    "                    else:\n",
    "                        read.qname = qname\n",
    "                    \n",
    "                    new_cigar = jcigar1\n",
    "                    if int(jumpInsertSize):\n",
    "                        new_cigar += jumpInsertSize + \"I\"\n",
    "                    if delta > 0:\n",
    "                        new_cigar += str(delta) + \"D\"\n",
    "                    \n",
    "                    if delta < MIN_GAP_LEN:\n",
    "                        jt += 400\n",
    "                    new_cigar += jcigar2\n",
    "                    read.cigarstring = new_cigar\n",
    "                    read.reference_start = ref1_start + int(jbegin1)\n",
    "                    if jt == 0:\n",
    "                        jt = 3\n",
    "                    read.tags = tags + [('JT', jt)]   \n",
    "                    read.tags += [('JS', int(jump_score))]\n",
    "                    jdelsize.append(delta)\n",
    "                    jjumpland.append(land)\n",
    "                    if not debug_output_debug_only or qname == debug_qname:    \n",
    "                        if not len(JT_INCLUDE) or jt in JT_INCLUDE:\n",
    "                            sam_file.write(read)  \n",
    "                        if jt == 3:\n",
    "                            jump_read_written += 1\n",
    "                    if qname == debug_qname: print(\"read.reference_start\", read.reference_start)\n",
    "                    if qname == debug_qname:\n",
    "                        print(\"jump new_cigar\", new_cigar)\n",
    "                            \n",
    "    \n",
    "            if debug: print(\"jump_read_written\", jump_read_written, \"jump_read_lowscore\", jump_read_lowscore, \"jump_read_lessthanref\", jump_read_lessthanref)\n",
    "            if len(jdelsize):\n",
    "                stats = [str(min(jdelsize)), str(max(jdelsize)), str(statistics.mean(jdelsize)), \"%.3f\" % min(jjumpland), \"%.3f\" % max(jjumpland), \"%.3f\" % statistics.mean(jjumpland)]\n",
    "            else:\n",
    "                stats = [\"0\" for x in range(6)]\n",
    "            if len(jump_better_delta):\n",
    "                stats2 = [str(min(jump_better_delta)), str(max(jump_better_delta)), str(statistics.mean(jump_better_delta))]\n",
    "            else:\n",
    "                stats2 = [\"0\" for x in range(3)]\n",
    "\n",
    "            if MODE == \"DEL\":\n",
    "                out_bed.write(line[:-1] + (\"\\t%d\\t%d\\t%d\\t%s\\n\" % (jump_read_written, jump_read_lowscore, jump_read_lessthanref, \"\\t\".join(stats))))\n",
    "            elif MODE == \"DUP\":\n",
    "                if DEBUG_DUP:\n",
    "                    out_bed.write(line[:-1] + (\"\\t%d\\t%d\\t%d\\t%s\\t%s\\n\" % (jump_better, jump_read_lowscore, jump_read_okscore, \"\\t\".join(stats2),str(jump_better_delta))))\n",
    "                else:\n",
    "                    out_bed.write(line[:-1] + (\"\\t%d\\t%d\\t%d\\n\" % (jump_better, jump_read_okscore, jump_read_lowscore)))\n",
    "            \n",
    "            if debug_break_after_one_region:\n",
    "                break\n",
    "            if debug_break_after_debug_seen and debug_seen:\n",
    "                break\n",
    "\n",
    "sam_file.close()\n",
    "\n",
    "# convert to sorted bam\n",
    "cmd = \"samtools sort \" + OUT_SAM + \" >\" + OUT_BAM\n",
    "print(\"cmd\", cmd)\n",
    "os.system(cmd)\n",
    "cmd = \"samtools index \" + OUT_BAM\n",
    "print(\"cmd\", cmd)\n",
    "os.system(cmd)\n",
    "print(\"nlines_total\", nlines_total)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
