{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "063ba889-7e02-4d63-84b6-4158f6e2b8bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cnv_realign - using detected cnv areas, realign cram file using jump aligner around cnv edges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "8b549c0c-9271-49cb-ba29-771e5bbb7f20",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['/Users/drorkessler/miniconda3/lib/python3.10/site-packages/ipykernel_launcher.py', '-f', '/Users/drorkessler/Library/Jupyter/runtime/kernel-71178e2b-e0de-4d42-9c55-bcaa02a6849f.json']\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'/Users/drorkessler/tmp/data/jump_align/251020/dups.bed'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# argumennt processing + debug environment files\n",
    "import sys\n",
    "import os\n",
    "\n",
    "MODE=\"DUP\"\n",
    "MATCH_SCORE = 2\n",
    "MISMATCH_SCORE = -8\n",
    "OPEN_SCORE = -18\n",
    "EXTEND_SCORE = -1\n",
    "JUMP_SCORE = 0\n",
    "\n",
    "MIN_MISMATCHES = 30\n",
    "SOFTCLIP_THRESHOLD = 30\n",
    "FETCH_READ_PADDING = 500\n",
    "FETCH_REF_PADDING = 0\n",
    "MIN_SEQ_LEN_JUMP_ALIGN_COMPONENT = 30\n",
    "MIN_GAP_LEN = 30\n",
    "MAX_READS_PER_CNV = 4000\n",
    "\n",
    "GCLOUD_AUTH = False\n",
    "INPUT_NAME_FULL = False\n",
    "\n",
    "print(sys.argv)\n",
    "tmp = \"/tmp/jump_align_input.\" + str(os.getpid())\n",
    "JUMP_ALIGN_CMD = [\"para_jalign\", str(MATCH_SCORE), str(MISMATCH_SCORE), str(OPEN_SCORE), str(EXTEND_SCORE), \"-1\", str(JUMP_SCORE), tmp]\n",
    "\n",
    "if \"dup_cnv_realign\" in sys.argv[0] or \"stdin\" in sys.argv[0]:\n",
    "    if len(sys.argv) < 5:\n",
    "        print(\"usage: \" + sys.argv[0] + \" <input-cram> <range-bed> <ref-fasta> <output-prefix> [<mode>] [<min-mismatches]\\n\")\n",
    "        sys.exit(-1)\n",
    "    # commandline invocation\n",
    "    IN_CRAM = sys.argv[1]\n",
    "    CNV_BED = sys.argv[2]\n",
    "    REF_FASTA = sys.argv[3]\n",
    "    OUT_SAM = sys.argv[4] + \".sam\"\n",
    "    if len(sys.argv) >= 6:\n",
    "        MODE = sys.argv[5]\n",
    "    if len(sys.argv) >= 7:\n",
    "        MIN_MISMATCHES = int(sys.argv[6])\n",
    "else:\n",
    "    \n",
    "    REF_FASTA = os.path.expanduser(\"~/tmp/ref/Homo_sapiens_assembly38.fasta\")\n",
    "    OUT_SAM = \"/tmp/cnv_realign.\" + str(os.getpid()) + \".sam\"\n",
    "\n",
    "    # DUP development\n",
    "    IN_CRAM = \"gs://ug-cromwell-tests/structural_variant/030945-NA24385-Z0114-CAACATACATCAGAT.cram\"\n",
    "    CNV_BED = os.path.expanduser(\"~/tmp/data/jump_align/251020/dups.bed\")\n",
    "\n",
    "    JUMP_ALIGN_CMD = [\"jump_align/\" + JUMP_ALIGN_CMD[0]] + JUMP_ALIGN_CMD[1:]\n",
    "\n",
    "    GCLOUD_AUTH = True\n",
    "    INPUT_NAME_FULL = True\n",
    "\n",
    "\n",
    "OUT_BED = OUT_SAM.replace(\".sam\", \".bed\")\n",
    "\n",
    "CNV_BED"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "1f2d6c23-4fd3-4f8e-a5fb-e6b8179b4d60",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCS_OAUTH_TOKEN set to: ya29.a0ATi...\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import subprocess\n",
    "\n",
    "if GCLOUD_AUTH:\n",
    "\n",
    "    # Run the gcloud command and capture its output\n",
    "    token = subprocess.check_output(\n",
    "        [\"gcloud\", \"auth\", \"print-access-token\"], text=True\n",
    "    ).strip()\n",
    "    \n",
    "    # Set the environment variable for the current process\n",
    "    os.environ[\"GCS_OAUTH_TOKEN\"] = token\n",
    "    \n",
    "    # Optional: print confirmation\n",
    "    print(\"GCS_OAUTH_TOKEN set to:\", token[:10] + \"...\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ad586376-9e10-4978-9a47-0432160f1e3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# open files\n",
    "import pysam\n",
    "reads_file = pysam.AlignmentFile(IN_CRAM, \"rb\", reference_filename=REF_FASTA)\n",
    "fasta_file = pysam.FastaFile(REF_FASTA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "5e6f5d70-4ee2-41d7-8af7-1e98aa5039d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "\n",
    "def run_process(command, input_path):\n",
    "  command \n",
    "  print(\"command\", \" \".join(command)) \n",
    "    \n",
    "  try:\n",
    "    process = subprocess.Popen(command, \n",
    "                              stdin=subprocess.PIPE, \n",
    "                              stdout=subprocess.PIPE, \n",
    "                              stderr=subprocess.PIPE, \n",
    "                              text=True) \n",
    "\n",
    "    stdout, stderr = process.communicate()\n",
    "    returncode = process.returncode\n",
    "    #print(stderr)\n",
    "\n",
    "  except subprocess.CalledProcessError as e:\n",
    "    print(f\"Error executing command: {e}\")\n",
    "    return None, e.returncode\n",
    "\n",
    "  return stdout, returncode"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "8904d4bc-a80d-4e5f-9356-9889cf619106",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "def count_md_mismatches(read):\n",
    "    \"\"\"\n",
    "    Count the number of mismatches in a pysam.AlignedSegment read using the MD tag.\n",
    "    Indels are not counted.\n",
    "    \"\"\"\n",
    "    try:\n",
    "        md_tag = read.get_tag(\"MD\")\n",
    "    except KeyError:\n",
    "        return None\n",
    "\n",
    "    # Find all letters in the MD string, which represent mismatches\n",
    "    mismatches = re.findall(r\"[A-Z]\", md_tag)\n",
    "    return len(mismatches)\n",
    "\n",
    "def count_nm_mismatches(read):\n",
    "    \"\"\"\n",
    "    Count the number of mismatches in a pysam.AlignedSegment read using the NM tag.\n",
    "    Indels are counted.\n",
    "    \"\"\"\n",
    "    try:\n",
    "        nm_tag = read.get_tag(\"NM\")\n",
    "    except KeyError:\n",
    "        return None\n",
    "\n",
    "    return int(nm_tag)\n",
    "\n",
    "def count_softclip_mismatches(read, reference):\n",
    "    \"\"\"\n",
    "    Count mismatches in soft-clipped regions (both left and right) of a read.\n",
    "    `read` is a pysam.AlignedSegment\n",
    "    `reference` is a pysam.FastaFile\n",
    "    \"\"\"\n",
    "    if read.is_unmapped:\n",
    "        return 0\n",
    "\n",
    "    seq = read.query_sequence\n",
    "    mismatches = 0\n",
    "    ref_name = read.reference_name\n",
    "    start = read.reference_start\n",
    "    end = read.reference_end\n",
    "\n",
    "    cigartuples = read.cigartuples\n",
    "    # CIGAR operations\n",
    "    SOFT_CLIP = 4\n",
    "\n",
    "    # Left soft clip\n",
    "    if cigartuples[0][0] == SOFT_CLIP:\n",
    "        clip_len = cigartuples[0][1]\n",
    "        clipped_bases = seq[:clip_len]\n",
    "        ref_start = max(0, start - clip_len)\n",
    "        ref_bases = reference.fetch(ref_name, ref_start, start)\n",
    "        for rb, qb in zip(ref_bases, clipped_bases):\n",
    "            if rb.upper() != qb.upper():\n",
    "                mismatches += 1\n",
    "\n",
    "    # Right soft clip\n",
    "    if cigartuples[-1][0] == SOFT_CLIP:\n",
    "        clip_len = cigartuples[-1][1]\n",
    "        clipped_bases = seq[-clip_len:]\n",
    "        ref_bases = reference.fetch(ref_name, end, end + clip_len)\n",
    "        for rb, qb in zip(ref_bases, clipped_bases):\n",
    "            if rb.upper() != qb.upper():\n",
    "                mismatches += 1\n",
    "\n",
    "    return mismatches\n",
    "\n",
    "md_values = []\n",
    "nm_values = []\n",
    "sc_values = []\n",
    "\n",
    "def accept_read(read):\n",
    "    if MIN_MISMATCHES <= 0:\n",
    "        return True\n",
    "    sc = count_softclip_mismatches(read, fasta_file);\n",
    "    nm = count_nm_mismatches(read)\n",
    "    return (sc + nm) >= MIN_MISMATCHES\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "9499a6cc-6b0d-45a4-82ca-7ff9c03b89a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import tempfile\n",
    "import random\n",
    "\n",
    "random.seed(0)\n",
    "\n",
    "# is read soft clipped\n",
    "def is_softclipped(read):\n",
    "    return read.cigartuples[0][0] == pysam.CSOFT_CLIP or read.cigartuples[-1][0] == pysam.CSOFT_CLIP\n",
    "\n",
    "def is_substential_softclipped(read):\n",
    "    return (read.cigartuples[0][0] == pysam.CSOFT_CLIP and read.cigartuples[0][1]) >= SOFTCLIP_THRESHOLD \\\n",
    "                or (read.cigartuples[-1][0] == pysam.CSOFT_CLIP and read.cigartuples[-1][1] >= SOFTCLIP_THRESHOLD)\n",
    "\n",
    "# process a single cnv\n",
    "def process_cnv(chrom, start, end, mode):\n",
    "\n",
    "    # get all reads that cross the two cnv edges\n",
    "    reads = dict()\n",
    "    reads_in_ref = [set(), set()]\n",
    "    refs = []\n",
    "    refs_extents = []\n",
    "    ref_id = 0\n",
    "    for loc in [start, end]:\n",
    "        rmin = max(0, loc - FETCH_READ_PADDING)\n",
    "        rmax = loc + FETCH_READ_PADDING\n",
    "        for read in reads_file.fetch(chrom, max(0, loc - FETCH_READ_PADDING), loc + FETCH_READ_PADDING):\n",
    "            if mode == \"DUP\" and (not is_substential_softclipped(read) and not accept_read(read)) :\n",
    "                continue\n",
    "            #if mode == \"DUP\" and read.is_supplementary:\n",
    "            #    continue\n",
    "            reads[read.qname] = read\n",
    "            rmin = min(rmin, read.reference_start)\n",
    "            rmax = max(rmax, read.reference_end)\n",
    "            reads_in_ref[ref_id].add(read.qname)\n",
    "        refs_extents.append([rmin, rmax])\n",
    "        ref_id += 1\n",
    "        \n",
    "    # extend references before and after\n",
    "    refs_extents[0][0] = max(0, refs_extents[0][0] - FETCH_REF_PADDING)\n",
    "    refs_extents[1][1] = refs_extents[1][1] + FETCH_REF_PADDING\n",
    "\n",
    "    # get references\n",
    "    for extents in refs_extents:\n",
    "        rmin, rmax = extents\n",
    "        ref = fasta_file.fetch(chrom, rmin, rmax)\n",
    "        refs.append([rmin, ref])\n",
    "\n",
    "    # create input file for jump aligner\n",
    "    ref_emitted = False\n",
    "    reads_in_order = []\n",
    "    subsample_ratio = 1.0\n",
    "    if len(reads) > MAX_READS_PER_CNV:\n",
    "        subsample_ratio = MAX_READS_PER_CNV / len(reads)\n",
    "        print(\"subsample_ratio\", subsample_ratio)\n",
    "    nlines = 0\n",
    "    jalign_input = tmp\n",
    "    if INPUT_NAME_FULL:\n",
    "        jalign_input += \"_\" + chrom + \":\" + str(start) + \"-\" + str(end)\n",
    "    with open(jalign_input, 'w') as f:\n",
    "        for read in reads.values():\n",
    "            if subsample_ratio < 1.0:\n",
    "                if random.random() > subsample_ratio:\n",
    "                    continue\n",
    "            if not accept_read(read):\n",
    "                continue\n",
    "            reads_in_order.append(read)\n",
    "            if not ref_emitted:\n",
    "                line = read.qname + \"\\t\" + read.seq + \"\\t\" + refs[1][1] + \"\\t\" + refs[0][1] + \"\\n\"\n",
    "                ref_emitted = True\n",
    "            else:\n",
    "                line = read.qname + \"\\t\" + read.seq + \"\\t=\\n\"\n",
    "            f.write(line)\n",
    "            nlines += 1\n",
    "\n",
    "    # run jump_align\n",
    "    JUMP_ALIGN_CMD[-1] = jalign_input\n",
    "    alignments = run_process(JUMP_ALIGN_CMD, jalign_input)\n",
    "    header_seen = False\n",
    "    realignments = []\n",
    "    rheader = []\n",
    "    for alignment, read in zip(alignments[0].split(\"\\n\"), [None, *reads_in_order]):\n",
    "        if not header_seen:\n",
    "            rheader = alignment.split(\"\\t\")\n",
    "            header_seen = True;\n",
    "        else:\n",
    "            a = alignment.split(\"\\t\")\n",
    "            in1 = read.qname in reads_in_ref[0]\n",
    "            in2 = read.qname in reads_in_ref[1]\n",
    "            realignments.append([read, refs[0][0], refs[1][0], a, in1, in2])\n",
    "    return (rheader, realignments, nlines)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "06cfec8e-78fa-41f6-9686-47d19a3ff551",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chr1:206052502-206058501\n",
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chr1:209904501-209920001\n",
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chr2:97194502-97204501\n",
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chr2:97208001-97234001\n",
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chr2:122526502-122537001\n",
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chr5:21478502-21497001\n",
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chr5:175931002-175943001\n",
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chr8:57203502-57215501\n",
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chr9:133060501-133081001\n",
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chr11:18920502-18944001\n",
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chr11:89956001-89982001\n",
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chr12:37709002-37733501\n",
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chr12:79761502-79769001\n",
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chr15:82440002-82455001\n",
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chr15:101752502-101765001\n",
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chr16:16597502-16607501\n",
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chr16:28605002-28615501\n",
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chr16:29443002-29466501\n",
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chr17:37890502-37910501\n",
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chr19:7019002-7032001\n",
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chr19:21563002-21580501\n",
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chr19:39889002-39895001\n",
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chr19:43324002-43329001\n",
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chr21:13753501-13766501\n",
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chr21:13871001-13899501\n",
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chr22:20332501-20342501\n",
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chrX:344502-365501\n",
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chrX:523002-568501\n",
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chrX:574502-605501\n",
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chrX:606502-636001\n",
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chrX:637002-667001\n",
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chrX:679002-692001\n",
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chrX:1038502-1049501\n",
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chrX:1149002-1165501\n",
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chrX:1237002-1275001\n",
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chrX:1276502-1314001\n",
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chrX:1403002-1408001\n",
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chrX:1589002-1637001\n",
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chrX:1735502-1763501\n",
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chrX:1795502-1806001\n",
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chrX:2244002-2277501\n",
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chrY:16167501-16179001\n",
      "command jump_align/para_jalign 2 -8 -18 -1 -1 0 /tmp/jump_align_input.91238_chrY:16269001-16316001\n",
      "nlines_total 1438\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'/tmp/cnv_realign.91238.bed'"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import statistics\n",
    "\n",
    "chrom_sizes = dict(zip(fasta_file.references, fasta_file.lengths))\n",
    "\n",
    "# loop on bed file, write output bed file\n",
    "nlines_total = 0\n",
    "with open(OUT_BED, \"w\") as out_bed:\n",
    "    with open(CNV_BED) as f:\n",
    "        for line in f:\n",
    "            if line.startswith(\"#\"):\n",
    "                continue\n",
    "            bed_line = line.strip().split()\n",
    "            bed_chrom, bed_start, bed_end = bed_line[:3]\n",
    "            bed_start = int(bed_start)\n",
    "            bed_end = int(bed_end)\n",
    "            #check for valid cnv\n",
    "            if bed_end + FETCH_READ_PADDING > chrom_sizes[bed_chrom]:\n",
    "                continue\n",
    "            rheader, realignments, nlines = process_cnv(bed_chrom, bed_start, bed_end, MODE)\n",
    "            nlines_total += nlines\n",
    "            jump_better = 0\n",
    "\n",
    "            for realignment in realignments:\n",
    "                in_ref = [False, False]\n",
    "                read, ref1_start, ref2_start, ainfo, in_ref[0], in_ref[1] = realignment\n",
    "    \n",
    "                # decode alignment info\n",
    "                qname, better, jscore, score1, score2, jgain, size1, size2 = ainfo\n",
    "                better = int(better)\n",
    "                size1 = int(size1)\n",
    "                size2 = int(size2)\n",
    "                if size1 < 20 or size2 < 20:\n",
    "                    continue\n",
    "                if better:\n",
    "                    jump_better += 1\n",
    "    \n",
    "            out_bed.write(line[:-1] + (\"\\t%d\\n\" % (jump_better)))\n",
    "            \n",
    "\n",
    "print(\"nlines_total\", nlines_total)\n",
    "OUT_BED"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25ea972f-ac7b-4416-864d-53cf4d91e612",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c27cd09-4036-424a-a754-4516ba83a63a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7abc7580-3e7d-4214-ab0a-eb08644c1b99",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
