{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "3a0c1472-ea15-4086-adf9-4da141e20b1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# pair_haps - mate pairs of haplotypes that overlap in their softclipped en\n",
    "# to optimize:\n",
    "# - no need to prefetch before and after parts\n",
    "# - pairing loop: right hap list traversal area can be limited\n",
    "# - find_best_match() can be optimized for the thresholds that are later enforced"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "57548aa1-716b-46f9-8fd6-2dcde565f5cd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['/Users/drorkessler/miniconda3/lib/python3.10/site-packages/ipykernel_launcher.py', '-f', '/Users/drorkessler/Library/Jupyter/runtime/kernel-d3d34487-26c4-432d-906f-0de57620f2ce.json']\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "# args\n",
    "print(sys.argv)\n",
    "if \"pair_haps\" in sys.argv[0] or \"stdin\" in sys.argv[0]:\n",
    "    if len(sys.argv) != 5:\n",
    "        print(\"usage: \" + sys.argv[0] + \" <input-cram> <in-bed> <ref-fasta> <output-prefix>\\n\")\n",
    "        sys.exit(-1)\n",
    "    # commandline invocation\n",
    "    IN_CRAM = sys.argv[1]\n",
    "    IN_BED = sys.argv[2]\n",
    "    REF_FASTA = sys.argv[3]\n",
    "    OUT_PREFIX = sys.argv[4]\n",
    "else:\n",
    "    IN_CRAM = os.path.expanduser(\"~/tmp/data/pair_haps/chr4_10M.bam\")\n",
    "    IN_BED = os.path.expanduser(\"~/tmp/data/pair_haps/chr4_ex1.bed\")\n",
    "    REF_FASTA = os.path.expanduser(\"~/tmp/ref/Homo_sapiens_assembly38.fasta\")\n",
    "    OUT_PREFIX = \"/tmp/pair_haps.\" + str(os.getpid())\n",
    "\n",
    "\n",
    "REALIGN=True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "0d20dc49-09b9-4109-97ad-559c703f6a71",
   "metadata": {},
   "outputs": [],
   "source": [
    "# collect breakpoints from read\n",
    "bp_quant = 500\n",
    "min_softclip = 60\n",
    "def quant(n):\n",
    "    return round((round(n / bp_quant) + 0.5) * bp_quant)\n",
    "    \n",
    "def collect_breakpoints(read):\n",
    "\n",
    "    lbp = []\n",
    "    rbp = []\n",
    "    \n",
    "    if read.cigartuples[-1][0] == pysam.CSOFT_CLIP and read.cigartuples[-1][1] >= min_softclip:\n",
    "        lbp.append((quant(read.reference_end), read))\n",
    "    if read.cigartuples[0][0] == pysam.CSOFT_CLIP and read.cigartuples[0][1] >= min_softclip:\n",
    "        rbp.append((quant(read.reference_start), read))\n",
    "\n",
    "    return (lbp, rbp)    \n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "42ccbe50-ca15-4824-aaaf-e379ee981b77",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bed_line ['chr4', '1', '10000000']\n",
      "len(lbp) 11479\n",
      "len(rbp) 12424\n"
     ]
    }
   ],
   "source": [
    "# loop on entries from bed file, process one at a time\n",
    "import pysam\n",
    "from itertools import groupby\n",
    "\n",
    "read_count = 0\n",
    "lbp = []\n",
    "rbp = []\n",
    "header = None\n",
    "with pysam.AlignmentFile(IN_CRAM, \"rb\", reference_filename=REF_FASTA) as samf:\n",
    "\n",
    "    header = samf.header\n",
    "    \n",
    "    # loop on bed regions\n",
    "    with open(IN_BED) as f:\n",
    "        for line in f:\n",
    "            bed_line = line.strip().split()\n",
    "            print(\"bed_line\", bed_line)\n",
    "            chrom, start, end = bed_line[:3]\n",
    "            start = int(start)\n",
    "            end = int(end)\n",
    "\n",
    "            # loop on reads\n",
    "            for read in samf.fetch(chrom, max(0, start - 1000), end + 1000):\n",
    "                read_count += 1\n",
    "                if read_count % 1000000 == 0:\n",
    "                    print(\"read_count\", read_count, read.reference_name + \":\" + str(read.reference_start))\n",
    "                    sys.stdout.flush()\n",
    "                    \n",
    "                bp = collect_breakpoints(read)\n",
    "                lbp += bp[0]\n",
    "                rbp += bp[1]\n",
    "\n",
    "# sort, groupby\n",
    "lbp.sort(key=lambda x:x[0])\n",
    "lbp = [(key, [x[1] for x in group]) for key, group in groupby(lbp, lambda x: x[0])]\n",
    "rbp.sort(key=lambda x:x[0])\n",
    "rbp = [(key, [x[1] for x in group]) for key, group in groupby(rbp, lambda x: x[0])]\n",
    "print(\"leftside breakpoints:\", len(lbp))\n",
    "print(\"rightside breakpoints:\", len(rbp))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "4448a63b-3653-46ec-810b-86f2e1d39b89",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "building pairs ...\n",
      "chr4:882250 (8.71%) ...\n",
      "chr4:1696750 (17.42%) ...\n",
      "chr4:2556250 (26.13%) ...\n",
      "chr4:3359250 (34.85%) ...\n",
      "chr4:4234250 (43.56%) ...\n",
      "chr4:5098750 (52.27%) ...\n",
      "chr4:5966250 (60.98%) ...\n",
      "chr4:6816750 (69.69%) ...\n",
      "chr4:7662250 (78.40%) ...\n",
      "chr4:8500750 (87.12%) ...\n",
      "chr4:9576750 (95.83%) ...\n",
      "number of pairs: 0\n"
     ]
    }
   ],
   "source": [
    "from difflib import SequenceMatcher\n",
    "\n",
    "# next we look for lr pairs which are 1K-10K in distance\n",
    "min_d = 1000\n",
    "max_d = 10000\n",
    "max_match_jump = 20\n",
    "min_match_size = 20\n",
    "portion_min = 0.5\n",
    "\n",
    "h1_debug = [\"HC_chr4:64571_1001\", \"HC_chr4:64571_1015\"]\n",
    "#h1_debug = [\"HC_chr4:41229_1002\", \"HC_chr4:64571_1015\"]\n",
    "h2_debug = [\"HC_chr4:67904_1008\", \"HC_chr4:67904_1006\"]\n",
    "on_debug = False\n",
    "find_longest_match_count = 0\n",
    "\n",
    "jumps_schema = (\"a\", \"b\", \"size\", \"p1\", \"p2\", \"h1[0].cigarstring\", \"h1[0].reference_start\", \"h1[0].reference_end\", \"len(h1[0].seq)\", \"h2[0].cigarstring\", \"h2[0].reference_start\", \"h2[0].reference_end\", \"len(h2[0].seq)\")\n",
    "\n",
    "\n",
    "def haps_are_paired(h1, h2):\n",
    "\n",
    "    after1 = h1[4]\n",
    "    after2 = h2[4]\n",
    "    result = False\n",
    "    jumps = None\n",
    "    global find_longest_match_count\n",
    "\n",
    "    a, b, size = SequenceMatcher(None, after1, after2).find_longest_match()\n",
    "    find_longest_match_count += 1\n",
    "    if on_debug and ((h1[0].qname in h1_debug and h2[0].qname in h2_debug)):\n",
    "        print(\"======\")\n",
    "        print(\"a\", a, \"b\", b, \"size\", size)\n",
    "    if size >= min_match_size and a <= max_match_jump and b <= max_match_jump:\n",
    "        p1 = size / len(after1)\n",
    "        p2 = size / len(after2)\n",
    "        if max(p1, p2) >= portion_min:\n",
    "            result = True;\n",
    "            jumps = (a, b, size, p1, p2, h1[0].cigarstring, h1[0].reference_start, h1[0].reference_end, len(h1[0].seq), h2[0].cigarstring, h2[0].reference_start, h2[0].reference_end, len(h2[0].seq))\n",
    "    \n",
    "    # debug\n",
    "    if on_debug and ((h1[0].qname in h1_debug and h2[0].qname in h2_debug)):\n",
    "        print(\"-----\")\n",
    "        print(\"h1\")\n",
    "        for h in h1[1:]:\n",
    "            print(\"-\", h)\n",
    "        print(\"h2\")\n",
    "        for h in h2[1:]:\n",
    "            print(\"-\", h)\n",
    "        print(\"result\", result, \"jumps\", jumps)\n",
    "\n",
    "    return (result, jumps)\n",
    "\n",
    "def match_haps(lhaps, rhaps):\n",
    "    h1s = []\n",
    "    h2s = []\n",
    "    matched = []\n",
    "    # extract sequences\n",
    "    for hap in lhaps:\n",
    "        clip = hap.cigartuples[-1][1]\n",
    "        before = hap.seq[:-clip]\n",
    "        after = hap.seq[-clip:]\n",
    "        h1s.append((hap, hap.qname, hap.cigartuples, before, after))\n",
    "    for hap in rhaps:\n",
    "        clip = hap.cigartuples[0][1]\n",
    "        before = hap.seq[:clip]\n",
    "        after = hap.seq[clip:]\n",
    "        h2s.append((hap, hap.qname, hap.cigartuples, before, after))\n",
    "\n",
    "    # look for matches\n",
    "    for h1 in h1s:\n",
    "        for h2 in h2s:\n",
    "            result, jump = haps_are_paired(h1, h2)\n",
    "            if result:\n",
    "                matched.append((h1[0], h2[0], h1[0].qname, h2[0].qname, jump))\n",
    "\n",
    "    return matched\n",
    "\n",
    "# this is naive\n",
    "pairs = []\n",
    "count = 0\n",
    "print(\"building pairs ...\")\n",
    "for l,lhaps in lbp:\n",
    "    count += 1\n",
    "    if count % 1000 == 0:\n",
    "        print(\"%s:%d (%.2f%%) ...\" % (lhaps[0].reference_name, l, count / len(lbp) * 100.0))\n",
    "    for r,rhaps in rbp:\n",
    "        if r-l >= min_d:\n",
    "            if r-l <= max_d:\n",
    "                pairs += match_haps(lhaps, rhaps)\n",
    "            else:\n",
    "                break\n",
    "    \n",
    "\n",
    "print(\"number of pairs:\", len(pairs))\n",
    "#for pair in pairs:\n",
    "#    print(pair)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "5cd77c02-dba2-4aa3-9452-9bec8230d37c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# group by left and right haps\n",
    "lgrp = [(x,list(g)) for x,g in groupby(sorted(pairs, key=lambda x: x[2]), lambda x: x[2])]\n",
    "rgrp = [(x,list(g)) for x,g in groupby(sorted(pairs, key=lambda x: x[3]), lambda x: x[3])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "8a58c00c-df5e-4c77-bf3a-eeb3ef485fc5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10\n"
     ]
    }
   ],
   "source": [
    "# sequence length of a cigar tuple\n",
    "def tuple_seq_len(t):\n",
    "    match t[0]:\n",
    "        case pysam.CMATCH | pysam.CINS | pysam.CINS | pysam.CEQUAL | pysam.CDIFF :\n",
    "            return t[1]\n",
    "    return 0\n",
    "\n",
    "#print(tuple_seq_len((pysam.CMATCH, 10)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "74cd1f81-5d82-4cba-bf4f-0401ac206113",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cigar_trim: tuples [(<CIGAR_OPS.CMATCH: 0>, 10)] seq_start 1 seq_len 2\n",
      "cigar_trim: after prefix trimming tuples [(<CIGAR_OPS.CMATCH: 0>, 9)]\n",
      "[(<CIGAR_OPS.CMATCH: 0>, 2)]\n",
      "cigar_trim: tuples [(<CIGAR_OPS.CINS: 1>, 5), (<CIGAR_OPS.CMATCH: 0>, 10)] seq_start 1 seq_len 20\n",
      "cigar_trim: after prefix trimming tuples [(<CIGAR_OPS.CINS: 1>, 4), (<CIGAR_OPS.CMATCH: 0>, 10)]\n",
      "[(<CIGAR_OPS.CINS: 1>, 4), (<CIGAR_OPS.CMATCH: 0>, 16)]\n"
     ]
    }
   ],
   "source": [
    "# cigar trimming\n",
    "def cigar_trim(tuples, seq_start, seq_len, debug = False):\n",
    "    out_tuples = []\n",
    "    seq_ofs = 0\n",
    "\n",
    "    if debug:\n",
    "        print(\"cigar_trim: tuples\", tuples, \"seq_start\", seq_start, \"seq_len\", seq_len)\n",
    "    \n",
    "    # trim cigar until we reach seq_start\n",
    "    while seq_ofs < seq_start and len(tuples):\n",
    "        t = tuples[0];\n",
    "        tail = tuples[1:]\n",
    "        t_len = tuple_seq_len(t)\n",
    "        seq_left = seq_start - seq_ofs\n",
    "        if t_len <= seq_left:\n",
    "            # total removal\n",
    "            seq_ofs += t_len\n",
    "            tuples = tail\n",
    "        else:\n",
    "            # split it \n",
    "            seq_ofs += seq_left\n",
    "            t = (t[0], t_len - seq_left)\n",
    "            tuples = [t] + tail\n",
    "\n",
    "    if debug:\n",
    "        print(\"cigar_trim: after prefix trimming tuples\", tuples)\n",
    "\n",
    "    # create result cigar until we reach seq_len\n",
    "    while seq_ofs < seq_start + seq_len and len(tuples):\n",
    "        t = tuples[0];\n",
    "        tail = tuples[1:]\n",
    "        t_len = tuple_seq_len(t)\n",
    "        seq_left = seq_start + seq_len - seq_ofs\n",
    "        if t_len <= seq_left:\n",
    "            # total copy\n",
    "            seq_ofs += t_len\n",
    "            tuples = tail\n",
    "            out_tuples += [t]\n",
    "        else:\n",
    "            # split it \n",
    "            seq_ofs += seq_left\n",
    "            out_t = t\n",
    "            t = (t[0], t_len - seq_left)\n",
    "            tuples = [t] + [tail]\n",
    "            out_t = (out_t[0], seq_left)\n",
    "            out_tuples += [out_t]\n",
    "\n",
    "    # if tuples did not fill seq_len, add M\n",
    "    if seq_ofs < seq_start + seq_len:\n",
    "        seq_left = seq_start + seq_len - seq_ofs\n",
    "        if out_tuples[-1][0] != pysam.CMATCH:\n",
    "            out_tuples += [(pysam.CMATCH, seq_left)]\n",
    "        else:\n",
    "            t = out_tuples[-1]\n",
    "            out_tuples = out_tuples[:-1] + [(pysam.CMATCH, t[1] + seq_left)]\n",
    "    return out_tuples\n",
    "        \n",
    "#print(cigar_trim([(pysam.CMATCH, 10)], 1, 2, True))\n",
    "#print(cigar_trim([(pysam.CINS, 5),(pysam.CMATCH, 10)], 1, 20, True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "7189bb33-e31c-48e5-866f-b967323104d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# realignment code\n",
    "def realign(grp, left):\n",
    "    debug = (grp[0] in h1_debug or grp[0] in h2_debug) and on_debug\n",
    "\n",
    "    if not REALIGN:\n",
    "        if left:\n",
    "            return grp[1][0][0]\n",
    "        else:\n",
    "            return grp[1][0][1]\n",
    "    \n",
    "    # sort pairs on lowest jump, get first\n",
    "    pair = sorted(grp[1], key=lambda x: max(x[4][0], x[4][1]))[0]\n",
    "    jump = pair[4]\n",
    "    if debug:\n",
    "        print(\"left\", left)\n",
    "        print(\"pair\", pair[2:4])\n",
    "        print(\"jump\", jump)\n",
    "\n",
    "    # create a copy, access other hap\n",
    "    if left:\n",
    "        d = pair[0].to_dict()\n",
    "        hap1 = pysam.AlignedSegment.from_dict(d, header)\n",
    "        hap2 = pair[1]\n",
    "    else:\n",
    "        hap1 = pair[0]\n",
    "        d = pair[1].to_dict()\n",
    "        hap2 = pysam.AlignedSegment.from_dict(d, header)\n",
    "\n",
    "    # modify alignment according to jump\n",
    "    if left:\n",
    "        del_delta = hap2.reference_start - hap1.reference_end + jump[1]\n",
    "        ins_delta = jump[0]\n",
    "        if debug:\n",
    "            print(\"del_delta\", del_delta, \"ins_delta\", ins_delta)\n",
    "        clip = hap1.cigartuples[-1][1]\n",
    "        hap1.cigartuples = hap1.cigartuples[:-1] # remove softclip\n",
    "        if ins_delta:\n",
    "            hap1.cigartuples += [(pysam.CDIFF, ins_delta)] # add X element\n",
    "        hap1.cigartuples += [(pysam.CDEL, del_delta - ins_delta)] # add D element\n",
    "        #hap1.cigartuples += [(pysam.CMATCH, clip - ins_delta)] # add M element for what was clipped\n",
    "        hap1.cigartuples += cigar_trim(hap2.cigartuples[1:], jump[1],  clip - ins_delta, False)\n",
    "        if debug:\n",
    "            print(\"new cigar: \", hap1.cigarstring)\n",
    "        hap1.tags += [('JO', hap2.qname), ('JP', str(jump))]\n",
    "\n",
    "        # manipulate ef and ez\n",
    "        ef = \"\"\n",
    "        ez = \"\"\n",
    "        for tag in hap2.tags:\n",
    "            if tag[0] == 'ef':\n",
    "                ef = tag[1]\n",
    "            elif tag[0] == 'ez':\n",
    "                ez = tag[1]\n",
    "        #print(\"ef\", ef)\n",
    "        #print(\"ez\", ez)\n",
    "        tags = []\n",
    "        for tag in hap1.tags:\n",
    "            if tag[0] == 'ef':\n",
    "                tag = ('ef', tag[1] +  \" \" + ef)\n",
    "            elif tag[0] == 'ez':\n",
    "                tag = ('ez', tag[1] + \" \" + ez)\n",
    "            tags += [tag]\n",
    "        hap1.tags = tags\n",
    "        \n",
    "        return hap1\n",
    "    else:\n",
    "        del_delta = hap2.reference_start - hap1.reference_end - 1 + jump[1]\n",
    "        ins_delta = jump[0]\n",
    "        #ins_delta = 0\n",
    "        if debug:\n",
    "            print(\"del_delta\", del_delta, \"ins_delta\", ins_delta)\n",
    "        clip = hap2.cigartuples[0][1]\n",
    "        hap2.cigartuples = hap2.cigartuples[1:] # remove softclip\n",
    "        hap2.cigartuples = [(pysam.CDEL, del_delta - ins_delta)] + hap2.cigartuples # add D element\n",
    "        if ins_delta:\n",
    "            hap2.cigartuples = [(pysam.CDIFF, ins_delta)] + hap2.cigartuples # add X element\n",
    "        hap2.cigartuples = [(pysam.CMATCH, clip - ins_delta)] + hap2.cigartuples # add M element for what was clipped\n",
    "        hap2.reference_start -= (del_delta + clip - ins_delta)\n",
    "        if debug:\n",
    "            print(\"new cigar: \", hap2.cigarstring)\n",
    "        hap2.tags += [('JO', hap1.qname), ('JP', str(jump))]\n",
    "\n",
    "        return hap2\n",
    "\n",
    "\n",
    "#realign(rgrp[0], False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "0d8e2671-9339-475e-8cff-bb481c84e004",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sam_fname /tmp/pair_haps.21251.sam\n",
      "regions_fname /tmp/pair_haps.21251_regions.bed\n",
      "cmd samtools sort /tmp/pair_haps.21251.sam >/tmp/pair_haps.21251.bam\n",
      "cmd samtools index /tmp/pair_haps.21251.bam\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create output file\n",
    "sam_fname = OUT_PREFIX + \".sam\"\n",
    "regions_fname = OUT_PREFIX + \"_regions.bed\"\n",
    "print(\"sam_fname\", sam_fname)\n",
    "print(\"regions_fname\", regions_fname)\n",
    "with pysam.AlignmentFile(sam_fname, \"w\", header=header) as sam_file:\n",
    "    with open(regions_fname, \"w\") as regions_file:\n",
    "\n",
    "        if 1:\n",
    "            for grp in lgrp:\n",
    "                read = realign(grp, True)\n",
    "                sam_file.write(read)\n",
    "\n",
    "        if 0:\n",
    "            for grp in rgrp:\n",
    "                read = realign(grp, False)\n",
    "                sam_file.write(read)\n",
    "\n",
    "        for pair in pairs:\n",
    "            h1 = pair[0]\n",
    "            h2 = pair[1]\n",
    "            regions_file.write(\"%s\\t%d\\t%d\\t%s::%s\\n\" % (h1.reference_name, h1.reference_start, h2.reference_end, h1.qname, h2.qname))\n",
    "            \n",
    "\n",
    "# convert to sorted bam\n",
    "bam_fname = OUT_PREFIX + \".bam\"\n",
    "cmd = \"samtools sort \" + sam_fname + \" >\" + bam_fname\n",
    "print(\"cmd\", cmd)\n",
    "os.system(cmd)\n",
    "cmd = \"samtools index \" + bam_fname\n",
    "print(\"cmd\", cmd)\n",
    "os.system(cmd)\n",
    "\n",
    "#print(\"-----\")\n",
    "#os.system(\"samtools view -c \" + bam_fname)\n",
    "#None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c41bdc1-21ca-486f-857f-7b3a68a6fdb7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3bf95a36-9f99-4f7d-b907-e9e94365da26",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4ad1fbc-adfe-4575-909f-e8e5d9736666",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "292208f6-9c52-4ae9-86f5-87f4174a805b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
